---
title: "ADAM10 Profile"
author: "Justin Bogias"
date: "`r format(Sys.Date(), '%d %b, %Y')`"
output:
  html_document: 
    toc: yes
    code_folding: hide
    toc_float: true
editor_options: 
  chunk_output_type: console
---

```{r}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warnings = FALSE)
```

With this workflow, I am creating a visualisation of the *ADAM10* transcriptomic profile. In this workflow, the gene expression, transcript expression, transcript proportions, and isoform structures are all plotted together to offer a holistic view of gene activity for *ADAM10*

# Load Packages
```{r message=FALSE, warning=FALSE}
library(rmarkdown)
library(rmdformats)
library(dplyr)
library(readr)
library(stringr)
library(tidyr)
library(tibble)
library(magrittr)
library(edgeR)
library(ggplot2)
library(tximport)
library(EnsDb.Hsapiens.v75)
library(wiggleplotr)
library(assertthat)
library(cowplot)
library(here)

source("/home/justin/Documents/R_Projects/Early_Gestation/markdowns/04_ADAM10_Plotting/R/plot_expression.R")
source("/home/justin/Documents/R_Projects/Early_Gestation/markdowns/04_ADAM10_Plotting/R/plot_proportions.R")
source("/home/justin/Documents/R_Projects/Early_Gestation/markdowns/04_ADAM10_Plotting/R/plot_structures.R")

# source(here("R/plot_expression.R"))
# source(here("R/plot_proportions.R"))
# source(here("R/plot_structures.R"))
```

# Load data
```{r}
dgelist <- read_rds(here("data/dgelist.rds"))
dtelist <- read_rds(here("data/dtelist.rds"))

dtelist$samples <- dtelist$samples %>%
  mutate(
    GestationGroup = ifelse(GestationalAge <= 10,
                            paste0("6", "\U2012", "10"),
                            paste0("11", "\U2012", "23")
    ),
    GestationGroup = factor(GestationGroup, levels = c(
      paste0("6", "\U2012", "10"), paste0("11", "\U2012", "23"))
    )
  )
```

# Visualisation of *ADAM10*
## Plot *ADAM10* gene expression
```{r gene_expression}
gene_plot <- plot_txp(
  dgelist = dgelist,
  target_gene = "ADAM10",
  variable = "GestationalAge",
  level = "gene",
  path = NULL,
  keep_legend = FALSE
) +
  labs(x = "Gestational Age (weeks)") +
  ggtitle(NULL) +
  theme(axis.title = element_text(size = 20, colour = "black"),
        axis.text = element_text(size = 16, colour = "black"))
```

## Plot *ADAM10* transcript expression
```{r transcript_expression}
transcript_plot <- plot_txp(
  dgelist = dtelist,
  target_gene = "ADAM10",
  variable = "GestationalAge",
  level = "transcript",
  tpm_method = "tximport",
  path = here("data/counts/salmon_abundance.csv.gz"),
  keep_legend = FALSE
) +
  labs(x = "Gestational Age (weeks)") +
  ggtitle(NULL) +
  theme(axis.title = element_text(size = 20, colour = "black"),
        axis.text = element_text(size = 16, colour = "black"))
```

## Plot *ADAM10* transcript proportions
```{r transcript_proportions}
anno <- data.frame(x1 = c(1, 0, 0, 1),
                   x2 = c(2, 0, 0, 2), 
                   y1 = c(0.78, 0, 0, 0.78),
                   y2 = c(0.84, 0, 0, 0.84), 
                   xstar = c(1.5, 0, 0, 1.5),
                   ystar = c(0.86, 0, 0, 0.86),
                   lab = c("*", "", "", "*"),
                   transcript_name = c("ADAM10-001",
                                       "ADAM10-002",
                                       "ADAM10-008",
                                       "ADAM10-015"))

proportion_plot <- plot_proportions(
  dtelist = dtelist,
  target_gene = "ADAM10",
  variable = "GestationGroup",
  tpm_method = "tximport",
  path = here("data/counts/salmon_abundance.csv.gz"),
  keep_legend = FALSE
) +
  labs(x = "Gestation Group (weeks)") +
  ggtitle(NULL) +
  theme_bw()

proportion_plot <- proportion_plot +
  scale_y_continuous(limits = c(0, 0.9)) +
  geom_text(data = anno,
            aes(x = xstar,
                y = ystar,
                label = lab),
            size = 7) +
  geom_segment(data = anno,
               aes(x = x1, 
                   xend = x1,
                   y = y1,
                   yend = y2),
               colour = "black") +
  geom_segment(data = anno,
               aes(x = x2,
                   xend = x2,
                   y = y1,
                   yend = y2)) +
  geom_segment(data = anno,
               aes(x = x1,
                   xend = x2, 
                   y = y2,
                   yend = y2),
               colour = "black") +
  theme(legend.position = "none",
        strip.background = element_rect(fill = "lightblue"),
        strip.text = element_text(colour = "black", size = 14),
        axis.title = element_text(size = 20, colour = "black"),
        axis.text = element_text(size = 14, colour = "black"))
```

## Visualise *ADAM10* isoform structures
```{r transcript_structures}
structure_plot <- plot_structures(
  dtelist = dtelist,
  target_gene = "ADAM10",
  annotation = EnsDb.Hsapiens.v75
) +
  scale_x_continuous(limits = c(0, 8000)) +
  theme(axis.title = element_text(size = 20, colour = "black"),
        axis.text = element_text(size = 16, colour = "black"))
```

## Create *ADAM10* profile plot

Prepare the cowplot and where each individual plot will be in it
```{r prepare_cowplot}
left_col <- cowplot::plot_grid(structure_plot,
                               proportion_plot,
                               labels = c("A", "C"),
                               nrow = 2)

right_col <- cowplot::plot_grid(gene_plot,
                                transcript_plot,
                                labels = c("B", "D"),
                                nrow = 2)

```

Now create the legend
```{r legend}
legend_plot <- cowplot::get_legend(
  plot_txp(
    dgelist = dtelist,
    target_gene = "ADAM10",
    variable = "GestationalAge",
    level = "transcript",
    path = here("data/counts/salmon_abundance.csv.gz"),
    keep_legend = TRUE
  ) +
    labs(colour = "Transcript", fill = "Transcript") +
    theme(legend.position = "right",
          legend.title = element_text(colour = "black",
                                      size = 20,
                                      face = "bold"),
          legend.text = element_text(colour = "black",
                                     size = 18))
)
```

...and add the legend
```{r add legend}
profile_plot <- cowplot::plot_grid(
  left_col,
  right_col,
  legend_plot,
  ncol = 3,
  rel_widths = c(1, 1, 0.3)
)
```

Prepare the title too
```{r title}
title_theme <- cowplot::ggdraw() +
  draw_label(
    paste0("ADAM10", " profile"),
    colour = "black",
    fontface = "bold",
    size = 35
  )
```

Now put the entire plot together
```{r ADAM10_cowplot}
adam10 <- cowplot::plot_grid(title_theme,
                             profile_plot,
                             nrow = 2,
                             rel_heights = c(0.05, 1))

ggsave(filename = "Figure_5.png",
       plot = adam10,
       device = "png",
       path = here("data/figures"),
       #       width = 41.25,
       #       height = 32.5875,
       width = 37.5,
       height = 29.625,
       units = "cm",
       dpi = 300)

#10.96 13.88

```

```{r}
adam10
```

# Functions Used in this Analysis

## plot_expression
```{r}
tpm <- function(dgelist) {
  rpk <- dgelist$counts/(dgelist$genes$length/1000)
  scaling <- colSums(rpk)/1e6
  tpm <- rpk/scaling
  tpm
}


plot_txp <- function(
    dgelist,
    target_gene,
    variable,
    level = "transcript",
    tpm_method = "standard",
    path = NULL,
    keep_legend = TRUE
) {
  
  require(edgeR)
  require(ggplot2)
  require(dplyr)
  require(readr)
  require(stringr)
  require(tidyr)
  require(tibble)
  require(magrittr)
  require(here)
  
  if (level == "transcript") {
    if (tpm_method == "tximport") {
      
      abundance_salmon <- read_csv(path)
      
      tpm <- abundance_salmon %>%
        as.data.frame() %>%
        set_rownames(abundance_salmon$feature_id) %>%
        dplyr::select(-feature_id) %>%
        as.matrix()
      
      lengthMat <- sapply(
        seq_len(ncol(tpm)),
        function(x) {dtelist$genes$length}
      ) %>%
        set_rownames(rownames(dtelist$counts)) %>%
        set_colnames(colnames(dtelist$counts))
      
      tx_counts <- tximport::makeCountsFromAbundance(
        countsMat = dtelist$counts,
        abundanceMat = tpm,
        lengthMat = lengthMat,
        countsFromAbundance = "lengthScaledTPM"
      )
      
      rna_counts <- tx_counts[rownames(dtelist$counts), ]
      
    } else if (tpm_method == "standard") {
      
      rna_counts <- tpm(dgelist)
      
    } else if (tpm_method == "none") {
      
      message("No tpm selected, assuming gene mode")
      
    } else {
      print("Either specify a path or just use standard")
      
    }
  } else if (level == "gene") {
    message("Running in gene mode, TPM not calculated...")
    rna_counts <- cpm(dgelist,
                      log = TRUE)
  }
  
  version_name_change <- dtelist$genes %>%
    dplyr::select("grch38_name" = gene_name,
                  "grch37_name" = transcript_name) %>%
    mutate(grch37_name = str_replace(grch37_name, "\\-.*", ""),
           same_name = grch38_name == grch37_name) %>%
    distinct()
  
  if(isFALSE(version_name_change %>%
             dplyr::filter(
               grch37_name == target_gene
             ) %>% .[["same_name"]]
  )) {
    target_gene_38 <- version_name_change %>%
      dplyr::filter(
        grch37_name == target_gene
      ) %>% .[["grch38_name"]]
  } else {
    target_gene_38 <- target_gene
  }
  
  metadata <- dgelist$samples
  gene_anno <- dgelist$genes
  
  gene_info <- gene_anno[gene_anno$gene_name %in% target_gene_38, ]
  
  rna_counts <- rownames_to_column(
    as.data.frame(rna_counts),
    if (level == "transcript") {
      var = "transcript_id"
    } else if (level == "gene") {
      var = "gene_id"
    }
  )
  
  rna_counts <- if (level == "transcript") {
    rna_counts[
      rna_counts$transcript_id %in% gene_info$transcript_id,
    ]
  } else if (level == "gene") {
    rna_counts[
      rna_counts$gene_id %in% gene_info$gene_id,
    ]
  }
  
  rna_ggplot <- pivot_longer(
    if (level == "transcript") {
      left_join(rna_counts,
                dplyr::select(gene_info,
                              "transcript_id",
                              "transcript_name"),
                by = "transcript_id")
    } else if (level == "gene") {
      left_join(rna_counts,
                dplyr::select(gene_info,
                              "gene_id",
                              "gene_name"),
                by = "gene_id")
    },
    cols = metadata$ID,
    names_to = "ID",
    values_to = "counts"
  )
  
  rna_ggplot <- left_join(
    rna_ggplot,
    dplyr::select(metadata,
                  "ID",
                  all_of(variable)),
    by = "ID"
  )
  
  if (level == "gene") {
    rna_ggplot <- rna_ggplot %>%
      left_join(version_name_change,
                by = c("gene_name" = "grch38_name")) %>%
      dplyr::rename("grch38_name" = "gene_name") %>%
      dplyr::rename("gene_name" = "grch37_name")
  } else {
    message("Transcript names are in GRCh37")
  }
  
  if (level == "transcript") {
    if (isTRUE(is.numeric(metadata[[variable]]))) {
      ggplot(rna_ggplot) +
        geom_point(
          aes(x = .data[[variable]],
              y = log2(counts),
              colour = transcript_name)
        ) +
        geom_smooth(
          aes(x = .data[[variable]],
              y = log2(counts),
              colour = transcript_name,
              fill = transcript_name)
        ) +
        scale_colour_viridis_d() +
        scale_fill_viridis_d() +
        expand_limits(y = 0) +
        xlab(variable) +
        ylab("Transcript Expression (log2 TPM)") +
        ggtitle(
          paste0(target_gene,
                 " transcript expression")
        ) +
        theme_bw() +
        theme(
          legend.title = element_blank()
        ) +
        if (isFALSE(keep_legend)) {
          theme(legend.position = "none")
        }
    } else {
      ggplot(rna_ggplot) +
        geom_boxplot(
          aes(x = .data[[variable]],
              y = log2(counts),
              fill = transcript_name),
          colour = "black"
        ) +
        scale_fill_viridis_d() +
        expand_limits(y = 0) +
        xlab(variable) +
        ylab("Transcript Expression (log2 TPM)") +
        ggtitle(
          paste0(target_gene, " transcript expression")
        ) +
        facet_grid(~ .data[[variable]],
                   scales = "free") +
        theme_bw() +
        theme(
          legend.title = element_blank()
        ) +
        if (isFALSE(keep_legend)) {
          theme(legend.position = "none")
        }
    }
    
  } else if (level == "gene") {
    if (isTRUE(is.numeric(metadata[[variable]]))) {
      ggplot(rna_ggplot) +
        geom_point(
          aes(x = .data[[variable]],
              y = counts,
              colour = gene_name)
        ) +
        geom_smooth(
          aes(x = .data[[variable]],
              y = counts,
              colour = gene_name,
              fill = gene_name)
        ) +
        scale_colour_viridis_d() +
        scale_fill_viridis_d() +
        expand_limits(y = 0) +
        xlab(variable) +
        ylab("Gene Expression (log2 CPM)") +
        ggtitle(
          paste0(target_gene, " gene expression")
        ) +
        theme_bw() +
        theme(
          legend.title = element_blank()
        ) +
        if (isFALSE(keep_legend)) {
          theme(legend.position = "none")
        }
    } else {
      ggplot(rna_ggplot) +
        geom_boxplot(
          aes(x = .data[[variable]],
              y = log2(counts),
              fill = gene_name),
          colour = "black"
        ) +
        scale_fill_viridis_d() +
        expand_limits(y = 0) +
        xlab(variable) +
        ylab("Gene Expression (log2 CPM)") +
        ggtitle(
          paste0(target_gene, " gene expression")
        ) +
        theme_bw() +
        theme(
          legend.title = element_blank()
        ) +
        if (isFALSE(keep_legend)) {
          theme(legend.position = "none")
        }
    }
  }
}

```

## plot_proportions
```{r}
plot_proportions <- function (
    dtelist,
    target_gene,
    variable,
    tpm_method = "standard",
    path = NULL,
    keep_legend = TRUE
) {
  
  require(ggplot2)
  require(dplyr)
  require(readr)
  require(stringr)
  require(tidyr)
  require(tibble)
  require(magrittr)
  require(tximport)
  require(here)
  
  if (tpm_method == "standard") {
    
    rpk <- dtelist$counts/(dtelist$genes$length/1000)
    scaling <- colSums(rpk)/1e6
    tpm <- rpk/scaling
    
  } else if (tpm_method == "tximport") {
    abundance_salmon <- read_csv(path)
    
    tpm <- abundance_salmon %>%
      as.data.frame() %>%
      set_rownames(abundance_salmon$feature_id) %>%
      dplyr::select(-feature_id) %>%
      as.matrix()
  } else {
    stop(
      "Need either abundance from 'tximport' or 'standard' method"
    )
  }
  
  version_name_change <- dtelist$genes %>%
    dplyr::select("grch38_name" = gene_name,
                  "grch37_name" = transcript_name) %>%
        mutate(grch37_name = str_replace(grch37_name, "\\-.*", ""),
               same_name = grch38_name == grch37_name) %>%
        distinct()
    
    if(isFALSE(version_name_change %>%
               dplyr::filter(
                   grch37_name == target_gene
               ) %>% .[["same_name"]]
    )) {
        target_gene_38 <- version_name_change %>%
            dplyr::filter(
                grch37_name == target_gene
            ) %>% .[["grch38_name"]]
    } else {
        target_gene_38 <- target_gene
    }

    gene_anno <- dtelist$genes

    metadata <- dtelist$samples

    metadata$variable <- metadata[[variable]]

    lengthMat <- sapply(seq_len(ncol(tpm)), function(x) {
        dtelist$genes$length
    })
    rownames(lengthMat) <- rownames(dtelist$counts)
    colnames(lengthMat) <- colnames(dtelist$counts)

    dtu_counts <- makeCountsFromAbundance(
        dtelist$counts,
        tpm,
        lengthMat,
        "lengthScaledTPM"
    )

    if (isTRUE(is.numeric(metadata[[variable]]))) {

        joined_data <- left_join(
            rownames_to_column(
                as.data.frame(
                    prop.table(
                        dtu_counts,
                        margin = 2
                    )
                ),
                var = "transcript_id"
            ),
            gene_anno,
            by = "transcript_id"
        )
        rownames(joined_data) <- joined_data$transcript_id

        prop2plot <- left_join(
            pivot_longer(
                rownames_to_column(
                    as.data.frame(
                        prop.table(
                            as.matrix(
                                joined_data[
                                    joined_data$gene_name %in% target_gene_38,
                                    colnames(joined_data) %in% metadata$ID
                                ]
                            ),
                            margin = 2
                        )
                    ), var = "transcript_id"
                ),
                cols = starts_with("PAC"),
                names_to = "ID",
                values_to = "Proportions"
            ),
            metadata,
            by = "ID"
        )

        prop2plot <- left_join(
            prop2plot,
            gene_anno[ c("transcript_id", "transcript_name") ],
            by = "transcript_id"
        )

        ggplot(
            prop2plot,
            aes(
                x = .data[[variable]],
                y = Proportions,
                fill = transcript_name
            )
        ) +
            geom_bar(
                position = "fill",
                stat = "identity",
                width = 0.95
            ) +
            ggtitle(
                paste0(
                    target_gene,
                    " transcript proportions"
                )
            ) +
            xlab(paste0(variable)) +
            ylab("Proportions") +
            scale_fill_viridis_d() +
            theme(
                panel.background = element_rect(
                    fill = "white",
                    colour = NA
                ),
                panel.grid = element_line(
                    colour = "white"
                )
            ) +
            if (isFALSE(keep_legend)) {
                theme(
                    legend.position = "none"
                )
            }

    } else {

        # Here we get the proportions of transcripts making up a gene
        box_props <- left_join(
            rownames_to_column(
                as.data.frame(
                    prop.table(
                        dtu_counts,
                        margin = 2
                    )
                ), var = "transcript_id"
            ),
            gene_anno[c("gene_id", "transcript_id", "gene_name")],
            by = "transcript_id"
        )

        box_props <- box_props[
            box_props$gene_name %in% target_gene_38,
        ]
        rownames(box_props) <- box_props$transcript_id

        # Here we calculate the proportions for the gene of interest
        # There is something in the way I plot these which cause the graphic to
        # change, I've check and found that the dataframe says the same thing
        # as the original one in the markdown
        box_props <- left_join(
            pivot_longer(
                rownames_to_column(
                    as.data.frame(
                        prop.table(
                            as.matrix(
                                box_props[metadata$ID]),
                            margin = 2)
                    ),
                    var = "transcript_id"),
                cols = starts_with("PAC"),
                names_to = "ID",
                values_to = "Proportions"),
            metadata,
            by = "ID"
        )

        box_props <- left_join(
            box_props,
            gene_anno[c("transcript_id", "transcript_name")],
            by = "transcript_id"
        )


        ggplot(box_props) +
            geom_boxplot(
                aes(
                    x = variable,
                    y = Proportions,
                    fill = transcript_name
                ),
                colour = "black"
            ) +
            ggtitle(
                paste0(
                    target_gene,
                    " transcript proportions"
                )
            ) +
            xlab(
                paste0(variable)
            ) +
            ylab(
                "Median transcript proportions"
            ) +
            scale_fill_viridis_d() +
            facet_grid(
                ~factor(transcript_name),
                scales = "free_x",
                space = "free",
                shrink = FALSE
            ) +
            theme_bw() +
            theme(
                axis.ticks.x = element_blank(),
                panel.background = element_rect(
                    fill = "white",
                    colour = NA
                ),
                panel.grid = element_line(
                    colour = "white"
                )
            ) +
            if (isFALSE(keep_legend)) {
                theme(
                    legend.position = "none"
                )
            }
    }
}
```

## plot_structures
```{r}
plot_structures <- function(
    dtelist,
    target_gene,
    annotation,
    rescale_introns = TRUE,
    new_intron_length = 50,
    flanking_length = c(50, 50),
    connect_exons = TRUE,
    transcript_label = TRUE,
    region_coords = NULL
) {
  
  require(rmarkdown)
  require(rmdformats)
  require(dplyr)
  require(readr)
  require(stringr)
  require(tidyr)
  require(tibble)
  require(magrittr)
  require(ggplot2)
  require(wiggleplotr)
  require(assertthat)
  
  version_name_change <- dtelist$genes %>%
    dplyr::select("grch38_name" = gene_name,
                  "grch37_name" = transcript_name) %>%
    mutate(grch37_name = str_replace(grch37_name, "\\-.*", ""),
           same_name = grch38_name == grch37_name) %>%
    distinct()
  
  if(isFALSE(version_name_change %>%
             dplyr::filter(
               grch37_name == target_gene
             ) %>% .[["same_name"]]
  )) {
    target_gene_38 <- version_name_change %>%
      dplyr::filter(
        grch37_name == target_gene
      ) %>% .[["grch38_name"]]
  } else {
    target_gene_38 <- target_gene
  }
  
  target_gene_37 <- target_gene
  
  tx_names <- dtelist$genes %>%
    dplyr::filter(gene_name == target_gene_38) %>%
    dplyr::select(transcript_id) %>%
    as.matrix() %>%
    as.character()
  
  tx_annot <- wiggleplotr:::extractTranscriptAnnotationsFromEnsembldb(
    annotation,
    target_gene_37,
    tx_names
  )
  
  exons <- tx_annot$exons
  cdss <- tx_annot$cdss
  transcript_annotations <- tx_annot$transcript_annotations
  
  if(is.null(cdss)){
    cdss = exons
  }
  
  # Check exons and cdss
  assertthat::assert_that(
    is.list(exons)|| is(exons, "GRangesList")
  )
  
  # Check that exons and cdss objects are lists
  assertthat::assert_that(
    is.list(cdss) || is(exons, "GRangesList")
  )
  
  # Join exons together
  joint_exons = wiggleplotr:::joinExons(exons)
  
  # Extract chromosome name
  chromosome_name = as.vector(
    GenomicRanges::seqnames(joint_exons)[1]
  )
  
  # If region_coords is specificed, then ignore the flanking_length attrbute and compute
  # flanking_length form region_coords
  if (!is.null(region_coords)) {
    gene_range = wiggleplotr:::constructGeneRange(
      joint_exons,
      c(0,0)
    )
    min_start = min(
      GenomicRanges::start(gene_range)
    )
    max_end = max(
      GenomicRanges::end(gene_range)
    )
    flanking_length = c(
      min_start - region_coords[1],
      region_coords[2] - max_end
    )
  }
  # Make sure that flanking_length is a vector of two elements
  assertthat::assert_that(
    length(flanking_length) == 2
  )
  
  # Rescale introns
  if (rescale_introns) {
    tx_annotations = wiggleplotr:::rescaleIntrons(
      exons,
      cdss,
      joint_exons,
      new_intron_length = new_intron_length,
      flanking_length
    )
    xlabel = "Distance from region start (bp)"
  } else {
    old_introns = wiggleplotr:::intronsFromJointExonRanges(
      GenomicRanges::ranges(
        joint_exons
      ),
      flanking_length = flanking_length
    )
    tx_annotations = list(
      exon_ranges = lapply(
        exons,
        GenomicRanges::ranges
      ),
      cds_ranges = lapply(
        cdss,
        GenomicRanges::ranges
      ),
      old_introns = old_introns,
      new_introns = old_introns
    )
    
    xlabel = paste(
      "Chromosome",
      chromosome_name,
      "position (bp)"
    )
  }
  
  # If transcript annotations are not supplied then construct them manually from the GRanges list
  if (is.null(transcript_annotations)) {
    plotting_annotations = dplyr::tibble(
      transcript_id = names(exons),
      strand = extractStrandsFromGrangesList(exons)
    ) %>%
      prepareTranscriptAnnotations()
  } else {
    plotting_annotations = wiggleplotr:::prepareTranscriptAnnotations(
      transcript_annotations
    )
  }
  
  # Plot transcript structures
  limits = c(
    min(
      IRanges::start(tx_annotations$new_introns)
    ),
    max(
      IRanges::end(tx_annotations$new_introns)
    )
  )
  structure = wiggleplotr:::prepareTranscriptStructureForPlotting(
    tx_annotations$exon_ranges,
    tx_annotations$cds_ranges,
    plotting_annotations
  )
  
  structure$transcript_label <- structure$transcript_label %>%
    gsub(
      pattern = ".*:",
      replacement = ""
    ) %>%
    gsub(
      pattern = " >",
      replacement = ""
    ) %>%
    as.data.frame() %>%
    set_colnames("IDs") %>%
    left_join(
      dtelist$genes,
      by = c("IDs" = "transcript_id")
    ) %>%
    dplyr::select("transcript_name") %>%
    as.matrix() %>%
    as.character()
  
  exons_df <- structure
  limits <- limits
  connect_exons <- connect_exons
  transcript_label <- transcript_label
  
  # Extract the position for plotting transcript name
  transcript_annot = dplyr::group_by(
    exons_df,
    transcript_id
  ) %>%
    dplyr::filter(
      feature_type == "exon"
    ) %>%
    dplyr::arrange(
      'transcript_id',
      'start'
    ) %>%
    dplyr::filter(
      row_number() == 1
    ) %>%
    mutate(
      colour_info = paste0(
        transcript_label,
        feature_type
      )
    )
  
  exons_df <- exons_df %>%
    mutate(
      colour_info = paste0(
        transcript_label,
        feature_type
      )
    )
  
  exons_df <- structure %>%
    dplyr::filter(
      feature_type == "exon"
    )
  
  cds_df <- structure %>%
    dplyr::filter(
      feature_type == "cds"
    )
  
  exons_df <- exons_df %>%
    mutate(
      transcript_label = replace_na(
        transcript_label,
        "LRG_151t1"
      )
    )
  
  transcript_annot <- transcript_annot %>%
    mutate(
      transcript_label = replace_na(
        transcript_label,
        "LRG_151t1"
      )
    )
  
  exons_df %>%
    ggplot() +
    geom_rect(
      aes(
        xmin = start,
        xmax = end,
        ymax = transcript_rank+0.25,
        ymin = transcript_rank-0.25,
        fill = transcript_label
      ),
      alpha = 0.5
    ) +
    geom_rect(
      aes(
        xmin = start,
        xmax = end,
        ymax = transcript_rank+0.25,
        ymin = transcript_rank-0.25,
        colour = transcript_label,
        fill = transcript_label
      ),
      data = cds_df
    ) +
    geom_line(
      aes(
        x = start,
        y = transcript_rank,
        group = transcript_rank,
        colour = transcript_label
      )
    ) +
    geom_line(
      aes(
        x = end,
        y = transcript_rank,
        group = transcript_rank,
        colour = transcript_label
      )
    ) +
    xlab(xlabel) +
    theme_bw() +
    theme(
      plot.margin = unit(
        c(0, 0, 0, 0),
        "line"
      ),
      axis.title.y = element_blank(),
      axis.text.y = element_blank(),
      axis.ticks.y = element_blank(),
      legend.position="none",
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      strip.text.y = element_text(
        colour = "black"
      ),
      strip.background = element_rect(
        fill = "grey85"
      )
    ) +
    scale_fill_viridis_d() +
    scale_colour_viridis_d() +
    geom_text(
      aes(
        x = start,
        y = transcript_rank + 0.30,
        label = transcript_label
      ),
      data = transcript_annot,
      hjust = 0,
      vjust = 0,
      size = 4
    )
}
```
