---
title: "Early Gestation DTU"
author: "Justin Bogias"
date: "`r format(Sys.Date(), '%d %b, %Y')`"
output:
  html_document:
    toc: yes
    code_folding: hide
    toc_float: true
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE
)
```

```{r loadPackage}
library(tidyverse)
library(ggrepel)
library(edgeR)
library(DRIMSeq)
library(stageR)
library(quantro)
library(qsmooth)
library(tximport)
library(magrittr)
library(ensembldb)
library(AnnotationHub)
library(DT)
library(here)
```

# Import annotations
Load in the ensembl gene annotations here to be used throughout the DGE analysis. This will connect to an ensembl annotation database in which the ensembl version 99 metadata will be stored as an object. This will give information on the gene ID and name, length, biotype and other information. This step may take some time as the data can be somewhat large.

```{r genesGR}
ah <- AnnotationHub()

ensDb <- ah[["AH78783"]]
genesGR <- ensembldb::genes(ensDb)
transGR <- transcripts(ensDb)
transGR$gene_name <- mcols(genesGR)[mcols(transGR)$gene_id,"gene_name"]
transGR$length <- lengthOf(ensDb, "tx")

ensDb <- ah[["AH10684"]]
transcript_names <- ensDb %>%
  as.data.frame() %>%
  dplyr::filter(type == "transcript") %>%
  dplyr::select("transcript_id",
                "transcript_name")
```

Despite the fact we are primarily only going to be using `ensembldb` for our annotations, we need the transcripts names to make the plots look a bit nicer. We can form a table with transcript IDs matched with their names here.
The code has been commented out as it takes time to run and I opted to leave it out of the analysis due to the biomaRt servers being somewhat unreliable and sometimes decide not to work.
As such, the transcript name key has been saved as a .csv file and can be loaded in for the analysis and the csv is provided.
```{r load_test_results}
oxygenGenes <- read_csv(here("data/results/oxygenGenes.csv.gz"))
oxygenTranscripts <- read_csv(here("data/results/oxygenTranscripts.csv.gz"))
```

# Data Importation

The sample metadata is loaded here, detailing the patient information. Many different variables are recorded in these datasets, but of primary concern to this analysis is gestational age. Two separate sets of metadata will be required which contain information of the data from the Pregnancy Advisory Centre (PAC). PAC samples cover all of early gestation, from 6-23 weeks gestation.

Some early first trimester placenta samples we not purely chorionic villous tissue. This most likely arose due to the difficulty in isolating a single tissue type from a placenta that is extremely small. Bulk placental tissue is likely to have been mixed in the samples.

```{r Patient_metadata}
impure_samples <- c(
  "PAC006", "PAC008", "PAC024", "PAC025",
  "PAC034", "PAC035", "PAC039", "PAC041",
  "PAC036", "PAC045", "PAC071", "PAC131"
  )
pd <- here("data/metadata/metadata.csv") %>%
  read_csv() %>%
  dplyr::filter(Cohort == "PAC") %>%
  mutate(Trimester = as.factor(Trimester),
         Timepoint = ifelse(GestationalAge <= 10, "6-10weeks", "11-23weeks"))
```

The counts data used for this study were sampled from the Pregnancy Advisory Centre (PAC) and contains 96 RNA-seq samples sourced from chorionic villous tissue from placenta at 6-23 weeks' gestation. The samples were sequenced on the Illumina HiSeq 2500 at an average sequencing depth of 35.8 million paired-end reads.
```{r transCounts}
transCounts <- read_rds(here("data/counts/salmon_object.rds"))

rownames(transCounts$counts) <-  gsub("\\..*", 
                                      "",
                                      rownames(transCounts$counts))

rownames(transCounts$annotation) <-  gsub("\\..*",
                                          "",
                                          rownames(transCounts$annotation))

colnames(transCounts$counts) <- colnames(transCounts$counts) %>%
  basename() %>%
  enframe(name = NULL) %>%
  inner_join(pd, by = c("value" = "filenames")) %>%
  .[["ID"]]

```

Import transcript abundances calculated from *salmon*. These counts are validated TPMs calculated from salmon. They differ slightly from TPMs calculated directly from the counts in that there are certain validation steps which were used to generate them. These are required to run the low-level functions within the high-level `tximport` function.
```{r load_abundance}
abundance_salmon <- read_csv(here("data/counts/salmon_abundance.csv.gz"))
```

# Data Processing
## Remove XY chromosomes

By plotting the ratio of male:female samples in our data at each gestational age, it is clear that genes on sex chromosomes may be overrepresented across each of the timepoints. Hence they are removed in the following code.
```{r autosomeTxpID}
autosomeTxpID <- transGR %>%
  subset(!seqnames %in% c("MT", "X", "Y")) %>%
  names() %>%
  intersect(rownames(transCounts$counts))

transCounts$counts <- transCounts$counts[autosomeTxpID,]
transCounts$annotation <- transCounts$annotation[autosomeTxpID,]
```

## Filtering and Normalisation

The *quantro* package can be used to test for global differences in distributions between groups to guide the choice of whether it is appropriate to use global normalization methods, such as quantile normalisation. For instance, if the variability between groups is larger than the variability within each group then there may be global differences present rendering quantile normalisation inappropriate (depending on the type and source of variation)

As described in the [quantro paper]("https://link.springer.com/article/10.1186/s13059-015-0679-0#") the main function performs two tests.
 * An ANOVA to test if the medians of distributions are different across groups. Differences across groups
could be attributed to unwanted technical variation (such as batch effects) or real global biological variation.
This is a helpful step for the user to verify if there is any technical variation unaccounted for.

* A test for global differences between the distributions across groups which returns a test statistic called
quantroStat. This test statistic is a ratio of two variances (similar to the idea of ANOVA): the variability of
the distributions within groups relative to the variability between groups. If the variability between groups is
sufficiently larger than the variability within groups, then this suggests global adjustment methods may not
be appropriate.

The reader is encouraged to look to the original paper for more information if desired.

A DTElist is created with the calculated normalisation factors using qsmooth. A common, but uncommonly addressed, question which arises in RNA-seq analyses is if the normalisation methods used in the workflow are appropriate based on the distribution of the reads in the data. There is a risk of normalising the biological differences in the data in certain scenarios which may be unnecessary, and at worst inappropriate, for the data under analysis. Most differential expression methods operate under the assumption that only a small proportion of the genes in question are differentially expressed and that differences in read distributions between testing samples are due to technical issues in the data.

The DTElist counts here will serve as the transcript counts to be implemented in the *tximport* `makeCountsFromAbundance` package
```{r DTElist}
filterCounts <- function(x, minCPM = 2, minSamp = 27) {
  cpm <- cpm(x)
  i <- rowSums(cpm > minCPM) > minSamp
  x[i,]
}

DTElist <- transCounts %>%
  with(counts / annotation$Overdispersion) %>%
  filterCounts() %>%
  .[,!colnames(.) %in% impure_samples] %>%
  DGEList(
    samples = colnames(.) %>%
      enframe(name = NULL, value = "ID") %>%
      left_join(pd),
    genes =  mcols(transGR)[rownames(.),] %>%
      as.data.frame() %>%
      dplyr::select(tx_id, tx_biotype, gene_id, gene_name, gc_content, length)
  )

trans_eset <- ExpressionSet(
  assayData = round(DTElist$counts, 0),
  phenoData = AnnotatedDataFrame(
    pd %>%
      as.data.frame() %>%
      set_rownames(.$ID) %>%
      .[colnames(DTElist$counts),]
    )
  )

keepMeID <- sapply(1:nrow(trans_eset), function(x){ any(exprs(trans_eset)[x,] != 0) })
trans_eset <- trans_eset[keepMeID,]

trans_qs <- qsmooth(exprs(trans_eset), group_factor = pData(trans_eset)$Timepoint)

DTElist$counts <- trans_qs@qsmoothData
```

Clean the matrix of salmon abundances needed for formation of the `tximport` counts. The commented-out code performs the steps required if I were to load in the alignment counts directly from the phoenix HPC. However, in this workflow, the data files have been downloaded from phoenix and these steps aren't needed
```{r salmon_abundance}
abundance_salmon <- abundance_salmon %>%
  as.data.frame() %>%
  set_rownames(abundance_salmon$feature_id) %>%
  dplyr::select(-feature_id) %>%
  as.matrix()
```

## Prepare tpm counts
Form a data frame of the dtu prepared counts so that it contains both the gene and transcript names
```{r length_matrix}
lengthMat <- sapply(seq_len(ncol(abundance_salmon)), function(x){
  DTElist$genes$length
  }) %>%
  set_rownames(rownames(DTElist$counts)) %>%
  set_colnames(colnames(DTElist$counts))
```

I don't think I can actually apply quantro here as I may upset the assumptions run in tximport. I maybe should instead try and manipulate the counts within the DRIMSeq object instead
This `makeCountsFromAbundance` function will automatically calculate the mean transcript lengths anyway
```{r dtu_counts}
dtu_counts <- tximport::makeCountsFromAbundance(
  countsMat = DTElist$counts,
  abundanceMat = abundance_salmon,
  lengthMat = lengthMat,
  countsFromAbundance = "lengthScaledTPM") %>%
  as.data.frame() %>%
  rownames_to_column(var = "feature_id") %>%
  left_join(as.data.frame(transGR) %>%
              dplyr::select("gene_id", "tx_id"),
            by = c("feature_id" = "tx_id")) %>%
  as_tibble()

dtu_counts %>%
  write_csv(here("data/counts/dtu_counts.csv.gz"))
```


# Differential Transcript Usage

## DRIMSeq analysis
The `dmDSdata` function required the metadata (and counts) to have specific column names by which it selects the appropriate fields of info
The top gene IDs for plotting can be selected by changing the argument in the `dplyr::slice` function. Specifying `1` simply means taking the top gene when ordered by the column specified in `dplyr::arrange`. Recommended columns for ordering are `pvalue` or `adj_pvalue`
```{r dmDSdata}
set.seed(123)

d <- dmDSdata(
  counts = as.data.frame(dtu_counts),
  samples = as.data.frame(
    pd %>%
      dplyr::filter(!ID %in% impure_samples) %>%
      dplyr::rename("sample_id" = "ID"))
)

inds <- 1:length(d@counts)
  mean_expression <- unlist(lapply(inds, function(g){
    mean(colSums(d@counts[[g]]), na.rm = TRUE)
  }))

  names(mean_expression) <- names(d@counts)

design_full <- model.matrix(~ Timepoint, data = DRIMSeq::samples(d))

dtu_results <- dmFilter(d,
    min_samps_gene_expr = 27,
    min_samps_feature_expr = 27,
    min_samps_feature_prop = 27,
    min_feature_prop = 0.1,
    min_gene_expr = 10,
    min_feature_expr = 10
    ) %>%
  dmPrecision(design = design_full) %>%
  dmFit(design = design_full) %>%
  dmTest(coef = "Timepoint11-23weeks")

write_rds(dtu_results, here("data/results/dtu_results.rds"))
```

## Plot precision
Now we have the gene-wise and common precision of our genes/features. It's similar to a dispersion value, but is more convenient to compute. The dispersion can be calculated by `dispersion = 1 / (1 + precision)`
We can also plot the log10 of precision by the log10 of mean expression and can see that as the number of features increase there is an increase in both mean expression and precision.

For genes of interest, you can make plots (bar plots, line plots, box plots, ribbon plots)
of observed and estimated transcript ratios from the Dirichlet-multinomial model. You have to
define the `group_variable` parameter which should indicate a variable from `samples(d)`.
Currently, plots can be done only for categorical variables. We choose the `group` column
since it corresponds to the comparison of our interest. Estimated proportions are marked
with diamond shapes. As an example, we plot the top significant gene.

Plotting the genewise precision against mean expression should show a positive relationship between the two. This is a nice plot to check the data
```{r Expr_vs_Precision}
dtu_results@mean_expression %>%
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_id") %>%
  dplyr::rename("meanExpr" = "V1") %>%
  dplyr::arrange(desc(meanExpr)) %>%
  left_join(dtu_results@genewise_precision %>%
              as.data.frame() %>%
              rownames_to_column(var = "gene_id"),
            by = "gene_id") %>%
  dplyr::rename("genewise_precision" = ".") %>%
  as_tibble() %>%
  ggplot(aes(meanExpr, genewise_precision)) +
  geom_point() +
  scale_x_log10() +
  scale_y_log10() +
  xlab("log10 gene expression") +
  ylab("log10 genewise precision") +
  theme_bw()
```

# Two-stage validation analysis

The two stage analysis is the one that *den Berge et al* (with *C Soneson* who authors pretty much everything you're using) proposed in their paper.

*DRIMSeq* has returned both gene and transcript level p-values which can be used as an input to the stage-wise analysis which is implemented in the *stageR* package. Using *stageR* is recommended by both the *DRIMSeq* and *stageR* authors as simply interpreting both gene-level and transcript-level adjusted p-values does not provide the appropriate FDR control necessary for confident results and applying a stage-wise testing method provides usefal biological intrepretation of these results and improved statistical perfromance.

Stage-wise testing consists of a screening stage and a confirmation stage. In the screening stage, gene-level BH-adjusted p-values are screening stage and a confirmation stage.

* In the screening stage, the gene-level BH-adjusted p-values are essentially screened to detect genes for which the hypothesis of interest is significantly rejected.

* In the confirmation stage, those genes which are retained from the screening stage are further considered. For each gene separately, transcript-level p-values arew adjusted to control the FWER and BH-adjusted significance level of the screening stage.

It is important to note that transcript-level stage-wise adjusted p-values for genes that do not pass the screening stage are set to NA. Also the stage-wise adjusted p-values can not be compared to a significance level other than chosen in the stage-wise analysis. If that is of interest, one has to rerun this analysis with the newly specified significance level.

Perform the stage_wise analysis. Calculate adjusted p values afterwards.
```{r StageRTx}
# Prepare the screening and validation objects
pScreen <- DRIMSeq::results(dtu_results)$pvalue %>%
  set_names(DRIMSeq::results(dtu_results)$gene_id)

pConfirmation <- matrix(DRIMSeq::results(
  dtu_results,
  level = "feature")$pvalue,
  ncol = 1
  ) %>%
  set_rownames(
    DRIMSeq::results(
      dtu_results,
      level = "feature")$feature_id
    )

tx2gene <- DRIMSeq::results(dtu_results,
                            level = "feature")[, c("feature_id", "gene_id")]

# Perform the two stage validation
oxygenDTU <- stageRTx(pScreen = pScreen,
                      pConfirmation = pConfirmation,
                      pScreenAdjusted = FALSE,
                      tx2gene = tx2gene) %>%
  stageWiseAdjustment(method = "dtu",
                      alpha = 0.05,
                      allowNA = TRUE) %>%
  getAdjustedPValues(order = TRUE,
                     onlySignificantGenes = FALSE) %>%
  dplyr::rename(
    Gene = geneID,
    Transcript = txID,
    geneFDR = gene,
    txpFDR = transcript
  ) %>%
  left_join(as.data.frame(genesGR),
            by = c("Gene" = "gene_id")
  ) %>%
  left_join(transcript_names, 
            by = c("Transcript" = "transcript_id")) %>%
  as_tibble()

oxygenDTU <- oxygenDTU %>%
  mutate("DTU" = geneFDR < 0.05)
```

Print DTU results
```{r}
sig_dtu <- oxygenDTU %>%
  dplyr::filter(geneFDR < 0.05)

print(dim(sig_dtu))
```

Save DTU results
```{r}
oxygenDTU %>% 
  write_csv(here("data/results/oxygenDTU.csv.gz"))

sig_dtu %>%
  write_csv(here("data/results/sig_dtu.csv.gz"))
```

# Data Visualisations
Visualisation of the number of isoforms per gene by using a pie chart. First, the number of isoforms in the entire DTU gene set of 2307 genes is plotted. 
```{r plotIsoforms}
table(oxygenDTU$gene_name) %>%
  table() %>%
  as.data.frame() %>%
  set_colnames(c("Isoforms", "Frequency")) %>%
  ggplot(
    aes(x = "",
        y = Frequency,
        fill = as.character(Isoforms))
    ) +
  geom_bar(stat = "Identity", width = 1) +
  coord_polar("y", start = 0) +
  scale_fill_viridis_d(begin = 1, end = 0) +
  ggtitle("Number of isoforms per gene in DTU") +
  labs(
    x = NULL,
    y = NULL,
    fill = NULL
    ) +
  theme_light() +
  theme(
    axis.line = element_blank(),
    axis.title.y = element_blank(),
    axis.ticks = element_blank()
    )
```

Generate table of significant DTU genes. The table is interactive and can be searched, sorted, etc.
```{r DTU_table}
oxygenDTU$DTU <- ifelse(oxygenDTU$geneFDR < 0.05, TRUE, FALSE)

oxygenDTU %>%
  dplyr::filter(DTU) %>%
  dplyr::select(
    gene_name,
    Gene,
    Transcript,
    start,
    end,
    geneFDR,
    txpFDR,
    gene_biotype,
    description
  ) %>%
  left_join(transcript_names,
            by = c("Transcript" = "transcript_id")
  ) %>%
  dplyr::rename(
    "gene-level FDR" = "geneFDR",
    "txp-level FDR" = "txpFDR",
    "Gene name" = "gene_name"
  ) %>%
  datatable(rownames = FALSE) %>%
  formatRound(
    columns = c("start", "end"),
    digits = 0
  ) %>%
  formatSignif(
    columns = c("gene-level FDR", "txp-level FDR"),
    digits = 3
  )
```

# Global Proportions
In DRIMseq, transcript proportions are calculated relative to their abundance compared against all other transcripts in a specific sample. This abundance is then aggregated into it's sample group and then compared against another experimental set. We can make our own changing proportion plot (deltaProp) to see how these changes are occuring and we may even be able to make a global representation of transcript proportion changes.


### Proportions per gene
In order to show the direction of proportion changes relative to each gene, we need to manually find the differences summarized to each gene
To start with, we need to calculate the proportions of transcripts per gene at each timepoint and then find the difference
```{r change_expr}
dtu_counts <- read_csv(here("data/counts/dtu_counts.csv.gz"))
dtu_results <- read_rds(here("data/results/dtu_results.rds"))

props <- dtu_counts %>%
  as.data.frame() %>%
  column_to_rownames("feature_id") %>%
  dplyr::select(-gene_id)

early_rownames <- props %>%
  dplyr::select(pd %>%
                  dplyr::filter(!ID %in% impure_samples) %>%
                  dplyr::filter(Timepoint == "6-10weeks") %>%
                  dplyr::select(ID) %>%
                  as.matrix() %>%
                  as.character) %>%
  rowMeans() %>%
  as.data.frame() %>%
  set_colnames("early") %>%
  rownames_to_column(var = "Transcript") %>%
  as_tibble()

change_expr <- props %>%
  dplyr::select(pd %>%
                  dplyr::filter(!ID %in% impure_samples) %>%
                  dplyr::filter(Timepoint == "11-23weeks") %>%
                  dplyr::select(ID) %>%
                  as.matrix() %>%
                  as.character) %>%
  rowMeans() %>%
  as.data.frame() %>%
  set_colnames("late") %>%
  rownames_to_column(var = "Transcript") %>%
  as_tibble() %>%
  left_join(early_rownames, by = "Transcript") %>%
  left_join(transGR %>% as.data.frame(),
            by = c("Transcript" = "tx_id")) %>%
  dplyr::select("Gene" = "gene_id",
                "Transcript",
                "early",
                "late")
```

Get proportion percentages in early gestation
```{r changing_props}
early_props <- tapply(change_expr$early %>%
                        set_names(change_expr$Transcript),
                      change_expr$Gene, prop.table) %>%
  unlist() %>%
  as.data.frame() %>%
  rownames_to_column(var = "IDs") %>%
  set_colnames(c("IDs", "proportions")) %>%
  mutate(IDs = str_replace_all(string = IDs, 
                               pattern = "[[:punct:]]",
                               replacement = "_and_"))

late_props <- tapply(change_expr$late %>% 
                        set_names(change_expr$Transcript),
                      change_expr$Gene, 
                      prop.table) %>%
  unlist() %>%
  as.data.frame() %>%
  rownames_to_column(var = "IDs") %>%
  set_colnames(c("IDs", "late_proportions")) %>%
  mutate(IDs = str_replace_all(string = IDs,
                               pattern = "[[:punct:]]",
                               replacement = "_and_")) %>%
  tidyr::separate(IDs,
                  c("Gene", "Transcript"),
                  "_and_") %>%
  dplyr::arrange(Gene)

changing_props <- early_props %>%
  tidyr::separate(IDs, c("Gene", "Transcript"), "_and_") %>%
  dplyr::arrange(Gene) %>%
  dplyr::select("Transcript", "early_prop" = "proportions") %>%
  left_join(late_props, by = "Transcript") %>%
  dplyr::select("Gene",
                "Transcript",
                "early_proportions" = "early_prop",
                "late_proportions") %>%
  mutate(deltaProp = (late_proportions - early_proportions))

# How many aren't changing proportions?
table(changing_props$deltaProp == 0)
```

Changing proportions for genes in the DTU analysis
```{r DTU_delta_props}
DTU_changing_proportions <- changing_props %>%
  dplyr::filter(Gene %in% oxygenDTU$Gene)

# Now check to see how many have no changes
table(DTU_changing_proportions$deltaProp == 0)
```

Changing proportions for genes exhibiting DTU
```{r sigDTU_delta_props}
DTU_changing_props <- changing_props %>%
  dplyr::filter(Gene %in% (oxygenDTU %>%
                  dplyr::filter(geneFDR < 0.05) %>%
                  dplyr::select(Gene) %>%
                  as.matrix() %>%
                  as.character()))
```

## Global Prop Figure 4
```{r plotProportions_txplevel}
global_prop_res <- dtu_results@mean_expression %>%
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  rownames_to_column(var = "gene_id") %>%
  set_colnames(c("gene_id", "meanExpr")) %>%
  dplyr::arrange(desc(meanExpr)) %>%
  left_join(
    dtu_results@results_feature,
    by = "gene_id"
  ) %>%
  dplyr::select(
    -"lr",
    -"df",
    -"pvalue",
    -"adj_pvalue"
  ) %>%
  left_join(oxygenGenes %>%
              dplyr::select(
                "Gene",
                "GeneName",
                "DEG_FDR" = "FDR"
              ),
            by = c("gene_id" = "Gene")) %>%
  left_join(oxygenTranscripts %>%
              dplyr::select(
                "Transcript",
                "DET_FDR" = "FDR"
              ),
            by = c("feature_id" = "Transcript")) %>%
  left_join(oxygenDTU %>%
             dplyr::select(
               "Transcript",
               "stageR_geneFDR" = "geneFDR",
               "stageR_txpFDR" = "txpFDR",
               "DTU"
             ),
            by = c("feature_id" = "Transcript")) %>%
  left_join(transcript_names,
            by = c("feature_id" = "transcript_id")) %>%
  left_join(DTU_changing_proportions %>%
              dplyr::select("Transcript",
                            "deltaProp"),
            by = c("feature_id" = "Transcript")) %>%
  mutate(signed_p = sign(deltaProp)*-log10(stageR_txpFDR)) %>%
  mutate(DTU = ifelse(stageR_txpFDR < 0.05, TRUE, FALSE)) %>%
  mutate(DEG = ifelse(DEG_FDR < 0.05, TRUE, FALSE)) %>%
  mutate(DET = ifelse(DET_FDR < 0.05, TRUE, FALSE)) %>%
  dplyr::select(
    -DET_FDR,
  ) %>%
  mutate(DTU = replace_na(DTU, FALSE)) %>%
  mutate(
    Status = case_when(
      DTU == TRUE & DET == FALSE & DEG == FALSE ~ "DTU",
      DTU == TRUE & DET == TRUE & DEG == FALSE ~ "DTU",
      DTU == TRUE & DET == TRUE & DEG == TRUE ~ "DTU_DEG",
      DTU == FALSE & DET == TRUE & DEG == TRUE ~ "DET_DEG",
      DTU == FALSE & DET == TRUE & DEG == FALSE ~ "DET",
      DTU == FALSE & DET == FALSE & DEG == TRUE ~ "DEG",
      DTU == TRUE & DET == FALSE & DEG == TRUE ~ "DTU",
      DTU == FALSE & DET == FALSE & DEG == FALSE ~ "Non Significant"
    )
  ) %>%
  mutate(
    Direction = case_when(
      signed_p > -log10(0.05) & abs(deltaProp) > 0.017 ~ "Up",
      signed_p < log10(0.05) & abs(deltaProp) > 0.017 ~ "Down",
      abs(deltaProp) < 0.017 ~ "NonSignif"
    )
  ) %>%
  replace_na(list(Status = "Non Significant",
                  Direction = "NonSignif")) %>%
  as_tibble()


global_prop_res_up <- global_prop_res %>%
  dplyr::filter(Direction == "Up")

global_prop_res_down <- global_prop_res %>%
  dplyr::filter(Direction == "Down")

global_prop_res_nonsig <- global_prop_res %>%
  dplyr::filter(Direction == "NonSignif")

p <- global_prop_res_nonsig %>%
  ggplot() +
  geom_point(
    aes(meanExpr,
        deltaProp),
    colour = "grey80",
    size = 3,
    alpha = 0.4
  ) +
  geom_point(
    data = global_prop_res_up,
    aes(meanExpr,
        deltaProp),
    size = 3, 
    alpha = 0.8,
    colour = "red"
  ) +
  geom_point(
    data = global_prop_res_down,
    aes(meanExpr,
        deltaProp),
    size = 3,
    alpha = 0.8,
    colour = "blue"
  ) +
  labs(x = "Average gene expression (log2 CPM)",
       y = "Difference in median proportions",
       title = "Difference in transcript proportions against transcript expression") +
  theme(
    legend.position = "none",
    panel.background = element_rect(fill = "white",
                                    colour = "grey"),
    panel.grid = element_line(colour = "grey92"),
    axis.text.x = element_text(angle = 0,
                               hjust = 1,
                               colour = "black",
                               size = 18),
    legend.title = element_text(size = 20,
                                colour = "black"),
    legend.text = element_text(size = 18,
                               colour = "black"),
    plot.title = element_text(size = 22,
                              colour = "black",
                              face = "bold"),
    axis.title = element_text(size = 20,
                              colour = "black"),
    axis.text.y = element_text(colour = "black",
                               size = 18)
  )
  
highlight1 <- dplyr::filter(
  global_prop_res,
  abs(deltaProp) > 0.2 & stageR_txpFDR < 0.05
)

highlight2 <- dplyr::filter(
  global_prop_res,
  meanExpr > 13 & stageR_txpFDR < 0.05 & abs(deltaProp) > 0.017
)

highlight3 <- dplyr::filter(
  global_prop_res,
  transcript_name == "ADAM10-001"
)

highlight4 <- dplyr::filter(
  global_prop_res,
  transcript_name == "GPR126-005"
)

highlight <- rbind(highlight1,
                   highlight2,
                   highlight3,
                   highlight4) %>%
  distinct() 

highlight <- highlight %>%
  mutate(Direction = factor(Direction,
                            levels = c("Up", "Down")))

highlight %<>% dplyr::filter(
  !transcript_name == "GPR126-005"
)

p + geom_label_repel(data = highlight,
                     aes(x = meanExpr,
                         y = deltaProp,
                         label = transcript_name,
                         colour = Direction),
                     show.legend = FALSE) +
  scale_colour_manual(values = c("red", "blue"))

```

