---
title: "Placental Transcription Profiling in Placentas at 6-23 weeks gestation reveals
  differential transcript usage"
author: "Justin Bogias"
date: "2024-07-31"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Load Packages
```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(magrittr)
library(edgeR)
library(DRIMSeq)
library(DESeq2)
library(stageR)
library(ggplot2)
library(cowplot)
library(AnnotationHub)
library(ensembldb)
library(ggfortify)
library(ggrepel)
library(DT)
library(quantro)
library(qsmooth)
library(VennDiagram)
library(goseq)
library(here)
```


# Annotations

First need to import annotations, the GRCh37 build of the human genome was used to align the RNA-seq data and so the GRCh37 annotations are used here.

For the annotations I elected to use the more recent annotations for gene and transcript length. Recent length annotations are more reliable than previous versions and they have been used in this analysis.

```{r annotations}
ah <- AnnotationHub()

ensDb <- ah[["AH78783"]]
genesGR <- ensembldb::genes(ensDb)
transGR <- transcripts(ensDb)
transGR$gene_name <- mcols(genesGR)[mcols(transGR)$gene_id,"gene_name"]
transGR$length <- lengthOf(ensDb, "tx")

tibble(x = 1:10, y = 2:11) %>% set_colnames(c("a", "b"))

ensDb <- ah[["AH10684"]]

transcript_names <- ensDb %>%
  as.data.frame() %>%
  dplyr::filter(type == "transcript") %>%
  dplyr::select("transcript_id",
                "transcript_name")

grch38_txp_anno <- transGR %>%
  as.data.frame() %>%
  dplyr::select("gene_id",
                "gene_name",
                "transcript_id" = "tx_id",
                "width",
                "length",
                "chromosome" = "seqnames",
                "transcript_biotype" = "tx_biotype") %>%
  left_join(transcript_names,
            by = "transcript_id") %>%
  as_tibble()

grch38_gene_anno <- genesGR %>%
  as.data.frame() %>%
  dplyr::select("gene_id",
                "gene_name",
                "length" = "width",
                "chromosome" = "seqnames",
                "gene_biotype") %>%
  as_tibble()
```


# PART ONE

# Data import

Now it's time to import the RNA-seq data. However, samples that were not pure chorionic villus tissue need to be removed from further analyses. Two other samples also had extremely poor sequencing quality and will also be removed. The samples are removed to mitigate any technical or systematic bias in the results.

```{r impure_samples}
impure_samples <- c(
  "PAC006", "PAC008", "PAC024", "PAC025",
  "PAC034", "PAC035", "PAC039", "PAC041", 
  "PAC036", "PAC045", "PAC071", "PAC131"
  )
```

Load in the patient metadata. The `Trimester` and `Timepoint` fields have been mutated into the metadata. `Trimester` defines whether the gestational age falls within the first (0 - 12 weeks) or second (13 - 26 weeks) trimesters of pregnancy.

```{r metadata}
pd <- here("data/metadata/metadata.csv") %>%
  read_csv() %>%
  dplyr::filter(Cohort == "PAC") %>%
  mutate(
    Trimester = as.factor(Trimester),
    Timepoint = ifelse(GestationalAge <= 10, "6-10weeks", "11-23weeks"),
    Timepoint = factor(Timepoint, levels = c("6-10weeks", "11-23weeks"))
  )
```

Load in the transcript counts returned from quantification with *salmon* after implementation of the Selective Alignment method. The counts were imported into R through the use of the `catchSalmon` function in `edgeR` prior to this analysis.

```{r counts}
transCounts <- readRDS(here("data/counts/salmon_object.rds")) 

rownames(transCounts$counts) <-  gsub(
  "\\..*",
  "",
  rownames(transCounts$counts)
)

rownames(transCounts$annotation) <-  gsub(
  "\\..*",
  "",
  rownames(transCounts$annotation)
)

colnames(transCounts$counts) <- colnames(transCounts$counts) %>%
  basename() %>%
  enframe(name = NULL) %>%
  inner_join(pd, 
             by = c("value" = "filenames")) %>%
  .[["ID"]]
```

Import transcript abundances calculated from *salmon*. These counts are validated TPMs calculated from salmon. They differ slightly from TPMs calculated directly from the counts in that there are certain validation steps which were used to generate them. They are useful for visualisation of transcripts in TPM

```{r load_abundance}
abundance_salmon <- read_csv(here("data/counts/grch37_PAC_abundance.csv.gz"))

abundance_salmon <- abundance_salmon %>%
  mutate(transcript_id = str_remove(transcript_id, "\\..*")) %>%
  column_to_rownames("transcript_id")

colnames(abundance_salmon) <- colnames(abundance_salmon) %>%
  basename() %>%
  enframe(name = NULL) %>%
  inner_join(pd, 
             by = c("value" = "filenames")) %>%
  .[["ID"]]

abundance_salmon <- abundance_salmon %>%
  rownames_to_column("transcript_id") %>%
  as_tibble()
```

# Data Preparation
By plotting the ratio of male:female samples in our data at each gestational age, it is clear that genes on sex chromosomes may be overrepresented across each of ghe timepoints. Hence they are removed in the following code.

```{r autosomeTxpID}
autosomeTxpID <- grch38_txp_anno %>%
  dplyr::filter(
    !chromosome %in% c("MT", "X", "Y")
    ) %>%
  as.matrix() %>%
  as.character() %>%
  intersect(
    rownames(transCounts$counts)
    )

transCounts$counts <- transCounts$counts[autosomeTxpID,]
transCounts$annotation <- transCounts$annotation[autosomeTxpID,]

abundance_salmon <- abundance_salmon %>% 
  dplyr::filter(
    transcript_id %in% autosomeTxpID
  )
```

# Create DGElist

```{r aggregate}
# aggregate the counts using simply aggregation -------------
txp_mat <- transCounts$counts

#gene_counts <- txp_mat %>%
  
gene_counts <- txp_mat %>%
  as.data.frame() %>% 
  rownames_to_column(
    var = "transcript_id"
  ) %>%
  left_join(
    grch38_txp_anno,
    by = "transcript_id"
  ) %>%
  dplyr::select(
    -"transcript_id"
  ) %>%
  group_by(gene_id) %>%
  summarise_if(
    is.numeric,
    sum
  ) %>%
  dplyr::select(
    -length,
    -width
  ) %>%
  dplyr::filter(
    !is.na(gene_id)
  ) %>%
  as.data.frame() %>%
  column_to_rownames("gene_id") %>%
  as.matrix() %>% 
  round(0)
```

```{r filter_counts_dge}
filterCounts <- function (
  x,
  minCPM = 2,
  minSamp = 27
) {
  cpm <- cpm(x)
  i <- rowSums(cpm > minCPM) > minSamp
  x[i,]
}
```

Now remove alternate chromosome builds
```{r removeHSGs_dge}
dge_counts <- gene_counts %>%
  as.data.frame() %>%
  rownames_to_column("gene_id") %>%
  dplyr::filter(str_detect(gene_id, "^E")) %>%
  as.data.frame() %>%
  column_to_rownames("gene_id") %>%
  as.matrix() %>%
  filterCounts() %>%
  .[,!colnames(.) %in% impure_samples]
```

Create DGElist
```{r DGElist}
DGElist <- dge_counts %>%
  DGEList(
    samples = colnames(.) %>%
      enframe(name = NULL, value = "ID") %>%
      left_join(pd),
    genes = rownames(.) %>%
      enframe(name = NULL, value = "gene_id") %>%
      left_join(grch38_gene_anno,
                by = "gene_id"))
```

# Smooth quantile normalisation

The *quantro* package can be used to test for global differences in distributions between groups to guide the choice of whether it is appropriate to use global normalization methods, such as quantile normalisation. For instance, if the variability between groups is larger than the variability within each group then there may be global differences present rendering quantile normalisation inappropriate (depending on the type and source of variation)

As described in the [quantro paper]("https://link.springer.com/article/10.1186/s13059-015-0679-0#") the main function performs two tests.
 * An ANOVA to test if the medians of distributions are different across groups. Differences across groups
could be attributed to unwanted technical variation (such as batch effects) or real global biological variation.
This is a helpful step for the user to verify if there is any technical variation unaccounted for.

* A test for global differences between the distributions across groups which returns a test statistic called
quantroStat. This test statistic is a ratio of two variances (similar to the idea of ANOVA): the variability of
the distributions within groups relative to the variability between groups. If the variability between groups is
sufficiently larger than the variability within groups, then this suggests global adjustment methods may not
be appropriate.

The reader is encouraged to look to the original paper for more information if desired.

```{r quantile_normalisation_dge}
eset <- ExpressionSet(
  assayData = round(DGElist$counts, 0),
  phenoData = AnnotatedDataFrame(
    pd %>%
      as.data.frame() %>%
      set_rownames(.$ID) %>%
      .[colnames(DGElist$counts),] 
    )
  )

keepMeID <- sapply(
  1:nrow(eset),
  function (x) { any(exprs(eset)[x,] != 0) }
)
eset <- eset[keepMeID,]

ds <- DESeqDataSetFromMatrix(
  countData = eset,
  colData = pData(eset),
  design = ~ Timepoint
)

r_ds <- vst(ds)
r_ds_mat <- as.matrix(assay(r_ds))

matdensity(
  r_ds_mat,
  groupFactor = pData(eset)$Timepoint,
  xlab = "counts",
  ylab = "density",
  main = "Gene counts",
  brewer.n = 8,
  brewer.name = "Dark2"
  )
legend(
  'top',
  c("NeuN_neg", "NeuN_pos"),
  col = c(1, 2),
  lty = 1,
  lwd = 3
)
```

The results of the quantro test
```{r quantro_dge}
qFit <- quantro(
  r_ds_mat, 
  groupFactor = pData(eset)$Timepoint,
  B = 1000
)
print(qFit)
```


Smooth quantile normalisation is applied as the differences between distributions aren't greatly exceeding sample variability. Smooth quantile normalisation is applied using methods from the [qsmooth paper](https://academic.oup.com/biostatistics/article/19/2/185/3949169#113055110)

```{r qsmooth_dge}
qs <- qsmooth(
  exprs(eset),
  group_factor = pData(eset)$Timepoint
)

DGElist$counts <- qs@qsmoothData
```


Transcripts were considered as detectbale if above `r formals(filterCounts)$minCPM`CPM in more than `r formals(filterCounts)$minSamp` samples.

```{r read_distributions_dge}
qsmooth::qsmoothData(qs) %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  pivot_longer(
    cols = starts_with("PAC"),
    names_to = "ID",
    values_to = "logCPM"
  ) %>%
  left_join(pd) %>%
  ggplot(
    aes(
      logCPM,
      stat(density), 
      colour = Timepoint, 
      group = ID
      )
  ) +
  geom_density() +
  scale_x_log10() +
  theme_bw()
```

# Principal Component Analysis
## Figure 1 - PCA
The samples are spread on the first principle component by gestational age
```{r PCA_dge}
DGElist$counts %>%
  cpm(log = TRUE) %>%
  t() %>%
  prcomp() %>%
  autoplot(
    data = DGElist$samples,
    colour = "GestationalAge",
    size = 5
  ) +
  ggtitle("Principal Component Analysis of Gene Expression") +
  labs(colour = "Gestational Age\n(weeks)") + 
  scale_colour_viridis_c(option = "magma") +
  theme_bw() +
  theme(axis.title = element_text(size = 14, 
                                  colour = "black"),
        axis.text = element_text(size = 12,
                                 colour = "black"),
        plot.title = element_text(size = 15,
                                  colour = "black",
                                  face = "bold")
  )
```


# Differential Expression Analysis
```{r set_parameters_dge}
alpha <- 0.05
lambda <- 1.4
```

For differential transcript expression we are testing to see if the expression of individual transcripts are exceeding a logFC between the before and after 10 weeks' samples as specified by $\lambda$ using the `glmQLFit` function.

null: logFC is betweem -$\lambda$ and +$\lambda$
alternate: absolute value of logFC is greater than $\lambda$

For all analyses for the value $\lambda$ the value `r lambda` was chosen

```{r oxygenGenes}
designMatrix <- model.matrix(
  ~ Timepoint +
    FetalSex,
  data = DGElist$samples
)

colnames(designMatrix) <- colnames(designMatrix) %>%
  str_remove("Timepoint") %>%
  str_remove("FetalSex")

DGElist <- DGElist %>%
  estimateCommonDisp(
    design = designMatrix
  )

oxygenGenes <- glmQLFit(DGElist,
                        design = designMatrix) %>%
  glmTreat(coef = "11-23weeks",
           lfc = log2(lambda)) %>%
  topTags(n = Inf) %>%
  .[["table"]] %>%
  as_tibble() %>%
  dplyr::mutate(DE = FDR < alpha & abs(logFC) > 1) %>%
  dplyr::rename("Gene" = "gene_id",
                "GeneName" = "gene_name",
                "Gene_biotype" = "gene_biotype")

sig_genes <- oxygenGenes %>%
  dplyr::filter(FDR < 0.05 & abs(logFC) > 1) 
```

Run a dimension on this for reproducibility
```{r print_gene_results}
print(dim(sig_genes))
```

Save results
```{r write_csv_genes}
oxygenGenes %>% 
  write_csv(here("data/results/oxygenGenes.csv.gz"))

sig_genes %>%
  write_csv(here("data/results/sig_genes.csv.gz"))
```


# Create DTElist
Create DTElist and calculate normalisation factors using qsmooth. A common, but uncommonly addressed, question which arises in RNA-seq analyses is if the normalisation methods used in the workflow are appropriate based on the distribution of the reads in the data. There is a risk of normalising the biological differences in the data in certain scenarios which may be unnecessary, and at worst inappropriate, for the data under analysis. Most differential expression methods operate under the assumption that only a small proportion of the genes in question are differentially expressed and that differences in read distributions between testing samples are due to technical issues in the data. 

```{r filter_counts_dte}
filterCounts <- function (
  x,
  minCPM = 2,
  minSamp = 27
) {
  cpm <- cpm(x)
  i <- rowSums(cpm > minCPM) > minSamp
  x[i,]
}
```

Now remove alternate chromosome builds
```{r removeHSGs_dte}
dte_counts <- transCounts %>%
 with(counts / annotation$Overdispersion) %>%
  as.data.frame() %>%
  rownames_to_column("transcript_id") %>%
  dplyr::filter(str_detect(transcript_id, "^E")) %>%
  as.data.frame() %>%
  column_to_rownames("transcript_id") %>%
  as.matrix() %>%
  filterCounts() %>%
  .[,!colnames(.) %in% impure_samples]
```

Create DTElist
```{r DTElist}
DTElist <- dte_counts %>%
  DGEList(
    samples = colnames(.) %>%
      enframe(name = NULL, value = "ID") %>%
      left_join(pd),
    genes = rownames(.) %>%
      enframe(name = NULL, value = "transcript_id") %>%
      left_join(grch38_txp_anno,
                by = "transcript_id"))
```


# Smooth quantile normalisation

The *quantro* package can be used to test for global differences in distributions between groups to guide the choice of whether it is appropriate to use global normalization methods, such as quantile normalisation. For instance, if the variability between groups is larger than the variability within each group then there may be global differences present rendering quantile normalisation inappropriate (depending on the type and source of variation)

As described in the [quantro paper]("https://link.springer.com/article/10.1186/s13059-015-0679-0#") the main function performs two tests.
 * An ANOVA to test if the medians of distributions are different across groups. Differences across groups
could be attributed to unwanted technical variation (such as batch effects) or real global biological variation.
This is a helpful step for the user to verify if there is any technical variation unaccounted for.

* A test for global differences between the distributions across groups which returns a test statistic called
quantroStat. This test statistic is a ratio of two variances (similar to the idea of ANOVA): the variability of
the distributions within groups relative to the variability between groups. If the variability between groups is
sufficiently larger than the variability within groups, then this suggests global adjustment methods may not
be appropriate.

The reader is encouraged to look to the original paper for more information if desired.

```{r quantile_normalisation_dte}
eset <- ExpressionSet(
  assayData = round(DTElist$counts, 0),
  phenoData = AnnotatedDataFrame(
    pd %>%
      as.data.frame() %>%
      set_rownames(.$ID) %>%
      .[colnames(DTElist$counts),]
    )
  )

keepMeID <- sapply(
  1:nrow(eset),
  function (x) { any(exprs(eset)[x,] != 0) }
)
eset <- eset[keepMeID,]

ds <- DESeqDataSetFromMatrix(
  countData = eset,
  colData = pData(eset),
  design = ~ Timepoint
)

r_ds <- vst(ds)
r_ds_mat <- as.matrix(assay(r_ds))

matdensity(
  r_ds_mat,
  groupFactor = pData(eset)$Timepoint,
  xlab = "counts",
  ylab = "density",
  main = "Transcript counts",
  brewer.n = 8,
  brewer.name = "Dark2"
  )
legend(
  'top',
  c("NeuN_neg", "NeuN_pos"),
  col = c(1, 2),
  lty = 1,
  lwd = 3
)
```

The results of the quantro test
```{r quantro_dte}
qFit <- quantro(
  r_ds_mat, 
  groupFactor = pData(eset)$Timepoint,
  B = 1000
)
print(qFit)
```


Smooth quantile normalisation is applied as the differences between distributions aren't greatly exceeding sample variability. Smooth quantile normalisation is applied using methods from the [qsmooth paper](https://academic.oup.com/biostatistics/article/19/2/185/3949169#113055110)

```{r qsmooth_dte}
qs <- qsmooth(
  exprs(eset),
  group_factor = pData(eset)$Timepoint
)

DTElist$counts <- qs@qsmoothData
```


Transcripts were considered as detectbale if above `r formals(filterCounts)$minCPM`CPM in more than `r formals(filterCounts)$minSamp` samples.

```{r read_distributions_dte}
qsmooth::qsmoothData(qs) %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  pivot_longer(
    cols = starts_with("PAC"),
    names_to = "ID",
    values_to = "logCPM"
  ) %>%
  left_join(pd) %>%
  ggplot(
    aes(
      logCPM,
      stat(density), 
      colour = Timepoint, 
      group = ID
      )
  ) +
  geom_density() +
  scale_x_log10() +
  theme_bw()
```


# Principal Component Analysis
## Figure 1 - PCA
The samples are spread on the first principle component by gestational age
```{r PCA_dte}
fig1 <- DTElist$counts %>%
  cpm(log = TRUE) %>%
  t() %>%
  prcomp() %>%
  autoplot(
    data = DTElist$samples,
    colour = "GestationalAge",
    size = 5
  ) +
  scale_colour_viridis_c(option = "magma") +
  ggtitle("Principal Component Analysis of Transcript Expression") +
  labs(colour = "Gestational Age\n(weeks)") + 
  theme_bw() +
  theme(axis.title = element_text(size = 20, 
                                  colour = "black"),
        legend.title = element_text(size = 20,
                                    colour = "black"),
        legend.text = element_text(size = 18,
                                   colour = "black"),
        axis.text = element_text(size = 18,
                                 colour = "black"),
        plot.title = element_text(size = 20,
                                  colour = "black",
                                  face = "bold")
  )

ggsave(filename = "Figure_1.png",
       plot = fig1,
       device = "png",
       path = here("data/figures"),
       height = 21.725,
       width = 27.5,
       units = "cm",
       dpi = 300)
```


# Differential Expression Analysis
```{r set_parameters_dte}
alpha <- 0.05
lambda <- 1.4
```

For differential transcript expression we are testing to see if the expression of individual transcripts are exceeding a logFC between the before and after 10 weeks' samples as specified by $\lambda$ using the `glmQLFit` function.

null: logFC is betweem -$\lambda$ and +$\lambda$
alternate: absolute value of logFC is greater than $\lambda$

For all analyses for the value $\lambda$ the value `r lambda` was chosen

```{r oxyTranscripts}
designMatrix <- model.matrix(
  ~ Timepoint +
    FetalSex,
  data = DTElist$samples
)

colnames(designMatrix) <- colnames(designMatrix) %>%
  str_remove("Timepoint") %>%
  str_remove("FetalSex")


DTElist <- DTElist %>%
  estimateCommonDisp(
    design = designMatrix
  )

oxygenTranscripts <- glmQLFit(DTElist,
                              design = designMatrix) %>%
  glmTreat(coef = "11-23weeks",
           lfc = log2(lambda)) %>%
  topTags(n = Inf) %>%
  .[["table"]] %>%
  as_tibble() %>%
  dplyr::mutate(DE = FDR < alpha & abs(logFC) > 1) %>%
  dplyr::rename("Transcript" = "transcript_id",
                "TranscriptName" = "transcript_name",
                "Transcript_biotype" = "transcript_biotype",
                "Gene" = "gene_id",
                "GeneName" = "gene_name")

sig_tx <- oxygenTranscripts %>%
  dplyr::filter(FDR < 0.05 & abs(logFC) > 1) 

```

Check dimensions here
```{r print_tx_results}
print(dim(sig_tx))
```


# PART TWO

```{r significance_data}
# metadata needs to be in this structure
age_order <- dplyr::arrange(pd, as.numeric(GestationalAge)) %>%
  dplyr::select(ID, GestationalAge)
```

# Figure 2 - Venn diagram
first we need to create subsets of the genes and transcripts which are considered  to be differentially expressed. With these subsetted lists we can overlap and record just how similar each one is.
```{r significant_genes}
sig_genes <- oxygenGenes %>%
  dplyr::filter(FDR < 0.05 & abs(logFC) > 1)

sig_tx <- oxygenTranscripts %>%
  dplyr::filter(FDR < 0.05 & abs(logFC) > 1)

sig_dtu <- oxygenDTU %>%
  dplyr::filter(geneFDR < 0.05)
```

Prepare objects to call in venn diagram, each object is essentially a list of genes which is featured in each different list. These recorded lists will be overlapped and the number of genes which feature in each of the different lists. Here, take the genes that transcribe significant transcripts in both the differential transcript expression and differential transcript usage results.
```{r DTE_DTU_genes}
DTE_DTU <- sig_dtu %>%
  dplyr::filter(Gene %in% sig_tx$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length()

sig_dtu %>%
  dplyr::filter(Gene %in% sig_tx$Gene,
                !Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length() %>%
  print()
```

Also identify here the transcripts that are differentially expressed but also are have a significant change in proportion. There are four genes that are significant in both DTE and DTU. This is because each gene has at least one transcript that has significantly changing proportions that is also differentially expressed. However, they also have another transcript with significant proportion changes but no significant differential expression.
```{r DTE_DTU_transcripts}
sig_dtu %>%
  dplyr::filter(Transcript %in% sig_tx$Transcript,
                !Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length() %>%
  print()

dte_dtu_gene <- sig_dtu %>%
  dplyr::filter(Gene %in% sig_tx$Gene,
                !Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct()

dte_dtu_tx <- sig_dtu %>%
  dplyr::filter(Transcript %in% sig_tx$Transcript,
                !Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct()

extra_genes <- dte_dtu_gene %>%
  dplyr::filter(!Gene %in% dte_dtu_tx$Gene)

sig_dtu %>%
  dplyr::filter(Gene %in% sig_tx$Gene,
                !Gene %in% sig_genes$Gene) %>%
  dplyr::filter(Gene %in% extra_genes$Gene) %>%
  dplyr::select("Gene",
                "Gene Name" = "gene_name",
                "Transcript",
                "Transcript Name" = "transcript_name",
                "DTU gene FDR" = "geneFDR",
                "DTU transcript FDR" = "txpFDR") %>%
  left_join(oxygenTranscripts %>%
              dplyr::select("Transcript", "DTE FDR" = "FDR"),
            by = "Transcript") %>%
  datatable()
```

Now check for genes that are DGE and DTU but not found in DTE
```{r DGE_DTU}
DGE_DTU <- sig_dtu %>%
  dplyr::filter(Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length()

sig_dtu %>%
  dplyr::filter(Gene %in% sig_genes$Gene,
                !Gene %in% sig_tx$Gene) %>%
  nrow() %>%
  print()
```

Now identify triply overlap genes
```{r DGE_DTE_DTU}
DGE_DTE_DTU <- sig_dtu %>%
  dplyr::filter(Gene %in% sig_tx$Gene,
                Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length()

sig_dtu %>%
  dplyr::filter(Gene %in% sig_tx$Gene,
                Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length() %>%
  print()
```

Now for genes in both DTE and DGE
```{r DTE_DGE}
DTE_DGE <- sig_genes %>%
  dplyr::filter(Gene %in% sig_tx$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length()

sig_genes %>%
  dplyr::filter(Gene %in% sig_tx$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length() %>%
  print()
```

Now get total numbers for the area parameters
```{r DGE_only}
DGE_number <- sig_genes$Gene %>%
  unique() %>%
  length()

DTE_number <- sig_tx$Gene %>%
  unique() %>%
  length()

DTU_number <- sig_dtu$Gene %>%
  unique() %>%
  length()
```

The objects created in the previous code chunck are used here to plot the triple venn diagram which records the level of overlap between each of the different analyses run.
```{r draw_venn}
grid.newpage()

v <- draw.triple.venn(
  area1 = DGE_number,
  area2 = DTE_number,
  area3 = DTU_number,
  n12 = DTE_DGE,
  n13 = DGE_DTU,
  n23 = DTE_DTU,
  n123 = DGE_DTE_DTU,
  category = c(
    "DGE",
    "DTE",
    "DTU"),
  fill = c(
    "blue", 
    "red",
    "green"),
  lty = "blank",
  cex = 2,
  fontfamily = "sans",
  cat.fontfamily = "sans",
  cat.pos = c(200, 160, 360),
  cat.dist = c(0.05, 0.06, 0.075),
  cat.cex = 2,
  scaled = FALSE
) 

grid.arrange(
  gTree(children = v),
  top = textGrob("Overlap of significant genes from DGE, DTE, and DTU analysis",
                 gp = gpar(fontsize = 22,
                           font = 2),
                 vjust = 1.5)
)

```

# Figure 3 - Greatest Differences
In this section I compare the DGE and DTE results in order to find which transcripts are most genuinly masked by aggregation to the gene level. Previously I had used the STAR counts from Jimmy's paper but have instead opted to aggregate my transcript counts to the gene level to see the effect of the going from the transcript-level to the gene-level
```{r DE_Differences}
DET_DEG_logFC_diff <- oxygenTranscripts %>%
  dplyr::select("Transcript",
                "Gene",
                "GeneName",
                "logCPM",
                "logFC",
                "FDR") %>%
  left_join(oxygenGenes %>%
              dplyr::select(
                "Gene",
                "gene_logFC" = "logFC",
                "geneFDR" = "FDR"
              ),
            by = "Gene") %>%
  left_join(transcript_names,
            by = c("Transcript" = "transcript_id")) %>%
  mutate(logFC_diff = (abs(logFC) - abs(gene_logFC))) %>%
  arrange(desc(logFC_diff))

DET_DEG_logFC_diff %>% dplyr::filter(logFC_diff > 1) %>%
  dplyr::filter(geneFDR > 0.05 & abs(gene_logFC) < 1) %>%
  dplyr::select(
    "Transcript ID" = "Transcript",
    "Transcript Name" = "transcript_name",
    "Gene Name" = "GeneName",
    "txp logFC" = "logFC",
    "logFC Diff" = "logFC_diff",
    "txp FDR" = "FDR",
    "gene FDR" = "geneFDR"
  ) %>%
  write_csv(here("data/results/top_difference.csv"))

significance_diff <- DET_DEG_logFC_diff %>% dplyr::filter(geneFDR > 0.05)

DET_DEG_logFC_diff %>%
  dplyr::filter(is.na(geneFDR)) %>%
  dplyr::select(
    "Transcript ID" = "Transcript",
    "Transcript Name" = "transcript_name",
    "Gene Name" = "GeneName",
    "logFC",
    "FDR"
  ) %>%
  write_csv(here("data/results/txp_only.csv"))
```

Can we do the test? If the logFCs are in the opposite direction, some differences in logFC would be telling us the wrong thing
```{r}
sig_tx %>%
  dplyr::select(
    "Gene_txp" = "Gene",
    "logFC_txp" = "logFC") %>%
  left_join(sig_genes %>%
              dplyr::select(
                "Gene_gene" = "Gene",
                "logFC_gene" = "logFC"
              ),
            by = c("Gene_txp" = "Gene_gene")) %>%
  ggplot(
    aes(logFC_txp,
        logFC_gene)
    ) +
  geom_point() +
  xlab("logFC transcripts") +
  ylab("logFC genes") +
  theme_bw() +
  theme(axis.title = element_text(size = 20),
        axis.text = element_text(size = 20))
```
Seems only CSH2 has a conflicting logFC 
Good to see I still see this even now!

# Compare FDR and logFC
```{r compare}
gene_tx_diff <- DET_DEG_logFC_diff %>%
  dplyr::filter(logFC_diff > 1) %>%
  dplyr::filter(geneFDR > 0.05 & abs(gene_logFC) < 1) %>%
  dplyr::select(
    "Transcript",
    "TranscriptName" = "transcript_name",
    "Gene",
    "GeneName",
    "txp_logFC" = "logFC",
    "gene_logFC",
    "logFC_diff",
    "txpFDR" = "FDR",
    "geneFDR"
  ) 
```


## Scale CPM for heatmap
```{r scaledCPM}
## Scale by rows
# In this function we are calculating the mean and standard deviation for every row of the
# matrix. Then we subtract the mean from the raw value and divide by the calculated standard
# deviation to generate our scaled values. Dividing by the standard deviation means that

scale_rows <- function(x){
  m = apply(x, 1, mean, na.rm = T)
  s = apply(x, 1, sd, na.rm = T)
  return((x - m) / s)
}

clusterOrder <- dtelist %>%
  cpm() %>%
  as.data.frame() %>%
  rownames_to_column(var = "Transcript") %>%
  dplyr::filter(Transcript %in% gene_tx_diff$Transcript) %>%
  dplyr::arrange(Transcript) %>%
  set_rownames(gene_tx_diff %>%
                 dplyr::filter(!Transcript == "ENST00000372728") %>%
                 dplyr::arrange(Transcript) %>%
                 dplyr::select(Transcript) %>%
                 as.matrix() %>%
                 as.character()) %>%
  dplyr::select(-Transcript) %>%
  as.matrix() %>%
  dist() %>%
  hclust() %>%
  as.dendrogram(method = "ward.D2") %>%
  order.dendrogram()

scaled_cpm <- dtelist %>%
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  rownames_to_column(var = "Transcript") %>%
  dplyr::filter(Transcript %in% gene_tx_diff$Transcript) %>%
  dplyr::arrange(Transcript) %>%
  set_rownames(gene_tx_diff %>%
                 dplyr::filter(!Transcript == "ENST00000372728") %>%
                 dplyr::arrange(Transcript) %>%
                 dplyr::select(Transcript) %>%
                 as.matrix() %>%
                 as.character()) %>%
  dplyr::select(-Transcript) %>%
  as.matrix()
```

## Plot pheatmap
```{r cluster_pheatmap}
clusteredMatrix <- scaled_cpm[clusterOrder, ]
clusteredMatrix <- clusteredMatrix[, (pd %>% 
                                        dplyr::filter(
                                          ID %in% colnames(clusteredMatrix)
                                        ) %>%
                                        dplyr::arrange(GestationalAge) %>%
                                        dplyr::select(ID) %>%
                                        as.matrix() %>%
                                        as.character)]

clusteredMatrix <- clusteredMatrix %>%
  as.data.frame() %>%
  rownames_to_column(var = "transcript_id") %>%
  left_join(transcript_names,
            by = "transcript_id") %>%
  dplyr::select(-transcript_id) %>%
  column_to_rownames("transcript_name") %>%
  as.matrix()

ann_col <- pd %>%
  dplyr::filter(ID %in% colnames(clusteredMatrix)) %>%
  dplyr::select(ID, "Gestation Group " = "Timepoint") %>%
  as.data.frame() %>%
  column_to_rownames("ID")

gene_row_order <- transcript_names %>%
  dplyr::filter(transcript_name %in% rownames(clusteredMatrix)) %>%
  left_join(oxygenTranscripts,
            by = c("transcript_id" = "Transcript")) %>%
  dplyr::select("Gene",
                "GeneName",
                "Transcript" = "transcript_id",
                "TranscriptName" = "transcript_name") %>%
  column_to_rownames("TranscriptName")

gene_row_order[clusterOrder, ]
```

Save figure 3
```{r}

pheatmap(clusteredMatrix,
         color = plasma(84),
         border_color = NA,
         show_colnames = FALSE,
         cluster_cols = FALSE,
         main = paste0(
           "Top 15 transcripts with greatest\n expression",
           " differences to the gene level"
         ),
         fontsize = 16,
         gaps_col = 27,
         cutree_rows = 4,
         clustering_method = "ward.D2",
         annotation_col = ann_col)

```

# Transcriptome profile

How many significant genes across all analyses?
```{r all_genes}
all_genes <- c(sig_tx$Gene, sig_genes$Gene, sig_dtu$Gene) %>%
  unique()

print(length(all_genes))
```

How many non coding in all 1642 genes?
```{r non_coding_genes}
grch38_gene_anno %>%
  dplyr::filter(gene_id %in% all_genes) %>%
  dplyr::count(gene_biotype, sort = TRUE) %>%
  print()
```

What are the lncRNAs?
```{r lncRNAs}
lncRNAs <- grch38_gene_anno %>%
  dplyr::filter(gene_id %in% all_genes) %>%
  dplyr::filter(gene_biotype == "lncRNA")

sig_genes %>%
  dplyr::filter(Gene %in% lncRNAs$gene_id) %>%
  print()

sig_tx %>%
  dplyr::filter(Gene %in% lncRNAs$gene_id) %>%
  print()

sig_dtu %>%
  dplyr::filter(Gene %in% lncRNAs$gene_id) %>%
  print()
```

What are all the significant transcripts between DTE and DTU?
```{r sig_txs}
all_tx <- c(sig_tx$Transcript, sig_dtu$Transcript) %>%
  unique()

print(length(all_tx))

all_tx_genes <- c(sig_tx$Gene, sig_dtu$Gene) %>%
  unique()

print(length(all_tx_genes))
```

How many isoforms did these genes have?
```{r isoforms}
oxygenTranscripts %>% 
  dplyr::select("Gene",
                "Transcript") %>%
  rbind(oxygenDTU %>%
          dplyr::select("Gene",
                        "Transcript")) %>%
  dplyr::filter(Gene %in% all_tx_genes) %>%
  distinct() %>%
  dplyr::count(Gene, sort = TRUE) %>%
  dplyr::count(n, sort = TRUE, name = "Number of Genes") %>%
  dplyr::rename("Isoforms" = "n") %>%
  print()
```

What types of transcripts were they?
```{r transcript_biotypes}
grch38_txp_anno %>%
  dplyr::filter(transcript_id %in% all_tx) %>%
  dplyr::count(transcript_biotype, sort = TRUE) %>%
  print()
```

Info on the triple overlap genes
```{r}
triple_genes <- sig_dtu %>%
  dplyr::filter(Gene %in% sig_tx$Gene,
                Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct()

oxygenTranscripts %>%
  dplyr::filter(Gene %in% triple_genes$Gene) %>%
  dplyr::select("Gene",
                "GeneName",
                "Transcript",
                "TranscriptName",
                "logFC",
                "FDR") %>%
  dplyr::count(GeneName, sort = TRUE)

oxygenDTU %>%
  dplyr::filter(Gene %in% triple_genes$Gene) %>%
  dplyr::select("Gene",
                "GeneName" = "gene_name",
                "Transcript",
                "TranscriptName" = "transcript_name",
                "geneFDR",
                "txpFDR") %>%
  dplyr::count(GeneName, sort = TRUE)

oxygenTranscripts %>%
  dplyr::filter(Gene %in% triple_genes$Gene) %>%
  dplyr::select("Gene",
                "GeneName",
                "Transcript",
                "TranscriptName",
                "logFC",
                "FDR") %>%
  dplyr::filter(FDR < 0.05 & abs(logFC) > 1) %>%
  dplyr::count(GeneName, sort = TRUE)

oxygenDTU %>%
  dplyr::filter(Gene %in% triple_genes$Gene) %>%
  dplyr::select("Gene",
                "GeneName" = "gene_name",
                "Transcript",
                "TranscriptName" = "transcript_name",
                "geneFDR",
                "txpFDR") %>%
  dplyr::filter(geneFDR < 0.05 & txpFDR < 0.05) %>%
  dplyr::count(GeneName, sort = TRUE) %>%
  dplyr::filter(n > 1)

oxygenDTU %>%
  dplyr::filter(Gene %in% triple_genes$Gene) %>%
  dplyr::select("Gene",
                "GeneName" = "gene_name",
                "Transcript",
                "TranscriptName" = "transcript_name",
                "geneFDR",
                "txpFDR") %>%
  dplyr::filter(geneFDR < 0.05 & txpFDR < 0.05) %>%
  dplyr::count(GeneName, sort = TRUE) %>%
  dplyr::filter(n == 1)
```

## What are their biotypes?
```{r}
oxt_tx_triple <- oxygenTranscripts %>%
  dplyr::filter(Gene %in% triple_genes$Gene)

oxy_dtu_triple <- oxygenDTU %>% 
  dplyr::filter(Gene %in% triple_genes$Gene)
```

Check to see we have complete representation of all transcripts here and then count the transcript biotypes
```{r}
all(oxy_dtu_triple$Transcript %in% oxt_tx_triple$Transcript)

oxt_tx_triple %>%
  dplyr::count(Transcript_biotype) %>%
  print()
```

What transcripts have retained introns?
```{r}
oxt_tx_triple %>% 
  dplyr::filter(Transcript_biotype == "retained_intron")
```

What transcripts have untranslated transcripts?
```{r}
oxt_tx_triple %>%
  dplyr::filter(Transcript_biotype == "processed_transcript")
```

Which transcript is targeted by nonsense mediated decay?
```{r}
oxt_tx_triple %>%
  dplyr::filter(Transcript_biotype == "nonsense_mediated_decay")
```


## How many isoforms in genes?
```{r}
sig_tx %>%
  dplyr::count(Gene, sort = TRUE) %>%
  print()

sig_tx %>% 
  dplyr::count(Gene,
               sort = TRUE, 
               name = "isoforms") %>%
  dplyr::count(isoforms,
               sort = TRUE) %>%
  print()
```

Which genes had the most amount of significantly DE isoforms?
```{r}
sig_tx %>%
  dplyr::count(Gene, sort = TRUE) %>%
  left_join(sig_tx %>%
              dplyr::select("Gene", "GeneName") %>%
              distinct(),
            by = "Gene") %>%
  print()
```

## Figure 3 transcripts information
```{r}
oxygenTranscripts %>%
  dplyr::filter(TranscriptName %in% rownames(clusteredMatrix))
```
