---
title: "Transcriptome Profile"
author: "Justin Bogias"
date: "`r format(Sys.Date(), '%d %b, %Y')`"
output: 
  rmdformats::readthedown
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warnings = FALSE)
```

```{r loadPackages, warning=FALSE, message=FALSE}
library(rmarkdown)
library(rmdformats)
library(dplyr)
library(readr)
library(stringr)
library(tidyr)
library(tibble)
library(magrittr)
library(ggplot2)
library(edgeR)
library(viridisLite)
library(pheatmap)
library(VennDiagram)
library(grid)
library(gridExtra)
library(DT)
library(here)
```

# Import annotations
Load in the ensembl gene annotations here to be used throughout the DTE analysis. This will connect to an ensembl annotation database in which the ensembl version 99 metadata will be stored as an object, giving information on the transcript ID, length, gc content, biotype and other information. This step may take some time as the data can be somewhat large.
```{r genesGR, eval=FALSE}
library(AnnotationHub)
library(ensembldb)

ah <- AnnotationHub()

ensDb <- ah[["AH78783"]]
genesGR <- genes(ensDb)
transGR <- transcripts(ensDb)
transGR$gene_name <- mcols(genesGR)[mcols(transGR)$gene_id,"gene_name"]
transGR$length <- lengthOf(ensDb, "tx")

ensDb <- ah[["AH10684"]]
transcript_names <- ensDb %>%
  as.data.frame() %>%
  dplyr::filter(type == "transcript") %>%
  dplyr::select("transcript_id",
                "transcript_name")

grch38_txp_anno <- transGR %>%
  as.data.frame() %>%
  dplyr::select("gene_id",
                "gene_name",
                "transcript_id" = "tx_id",
                "width",
                "length",
                "chromosome" = "seqnames",
                "transcript_biotype" = "tx_biotype") %>%
  left_join(transcript_names,
            by = "transcript_id") %>%
  as_tibble()

write_csv(grch38_txp_anno, here("data/grch38_tx_anno.csv.gz"))

grch38_gene_anno <- genesGR %>%
  as.data.frame() %>%
  dplyr::select("gene_id",
                "gene_name",
                "length" = "width",
                "chromosome" = "seqnames",
                "gene_biotype") %>%
  as_tibble()

write_csv(grch38_gene_anno, here("data/grch38_gene_anno.csv.gz"))

ensDb <- ah[["AH10684"]]
transcript_names <- ensDb %>%
  as.data.frame() %>%
  dplyr::filter(type == "transcript") %>%
  dplyr::select("transcript_id",
                "transcript_name")

write_csv(transcript_names, here("data/grch38_transcript_names.csv.gz"))
```

## Read in annotation data
Just read it in here because its quicker
```{r}
grch38_txp_anno <- read_csv(here("data/grch38_tx_anno.csv.gz"))
grch38_gene_anno <- read_csv(here("data/grch38_gene_anno.csv.gz"))
transcript_names <- read_csv(here("data/grch38_transcript_names.csv.gz"))
```

# Load data
Load the results from each analysis
```{r significance_data}
oxygenGenes <- read_csv(here("data/oxygenGenes.csv.gz"))
oxygenTranscripts <- read_csv(here("data/oxygenTranscripts.csv.gz"))
oxygenDTU <- read_csv(here("data/oxygenDTU.csv.gz"))

dtelist <- read_rds(here("data/dtelist.rds"))

impure_samples <- c(
  "PAC006", "PAC008", "PAC024", "PAC025",
  "PAC034", "PAC035", "PAC039", "PAC041", 
  "PAC036", "PAC045", "PAC071", "PAC131"
  )

pd <- here("data/metadata.csv") %>%
  read_csv() %>%
  dplyr::filter(Cohort == "PAC") %>%
  mutate(Trimester = as.factor(Trimester),
         Timepoint = ifelse(GestationalAge <= 10, "6-10weeks", "11-23weeks")) %>%
  dplyr::filter(!ID %in% impure_samples) %>%
  arrange(ID)

# metadata needs to be in this structure
age_order <- dplyr::arrange(pd, as.numeric(GestationalAge)) %>%
  dplyr::select(ID, GestationalAge)
```

# Differences Between Methods
I'm interested in finding how DGE, DTE, and DTU differ in their results. This isn't so much a technical goal, but a biological one. Primarily, I am looking for individual transcripts that are either differentially expressed or undergoing changes in proportion while no significant gene expression is seen for their parent gene. This goes for genes undergoing DTU as well. This would mean that changes in transcript expression remain undetected in gene-level analysis. Now I'm not here to talk trash about gene-level analyses, which is why I am equally as interested in the identification of significant genes with no significant transcripts that are differentially expressed or in different proportions. In this case, DGE analysis serves a wonderful purpose in the identification of potentially biologically significant changes in a gene's output that can not be detected in DTE.

First we need to create subsets of the genes and transcripts which are considered  to be differentially expressed. With these subsetted lists we can overlap and record just how similar each one is.
For the numbers identified in the venn diagram, we are using **only genes**, this is because comparing numbers of genes in DGE, and numbers of transcripts in DTE and DTU would give results showing more statistically significant features identified in DTE and DTU, when that is just because there are more transcripts than genes. It would be very misleading. Also we simply would fail to overlap the results. Basically the question here is: what genes do I identify in each analysis and where do these overlap?
```{r significant_genes}
sig_genes <- oxygenGenes %>%
  dplyr::filter(FDR < 0.05 & abs(logFC) > 1)

sig_tx <- oxygenTranscripts %>%
  dplyr::filter(FDR < 0.05 & abs(logFC) > 1)

sig_dtu <- oxygenDTU %>%
  dplyr::filter(geneFDR < 0.05)
```

Prepare objects to call in venn diagram, each object is essentially a list of genes which is featured in each different list. These recorded lists will be overlapped and the number of genes which feature in each of the different lists. Here, take the genes that transcribe significant transcripts in both the differential transcript expression and differential transcript usage results.
```{r DTE_DTU_genes}
DTE_DTU <- sig_dtu %>%
  dplyr::filter(Gene %in% sig_tx$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length()

sig_dtu %>%
  dplyr::filter(Gene %in% sig_tx$Gene,
                !Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length() %>%
  print()
```

Also identify here the transcripts that are differentially expressed but also are have a significant change in proportion. There are four genes that are significant in both DTE and DTU. This is because each gene has at least one transcript that has significantly changing proportions that is also differentially expressed. However, they also have another transcript with significant proportion changes but no significant differential expression.
```{r DTE_DTU_transcripts}
sig_dtu %>%
  dplyr::filter(Transcript %in% sig_tx$Transcript,
                !Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length() %>%
  print()

dte_dtu_gene <- sig_dtu %>%
  dplyr::filter(Gene %in% sig_tx$Gene,
                !Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct()

dte_dtu_tx <- sig_dtu %>%
  dplyr::filter(Transcript %in% sig_tx$Transcript,
                !Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct()

extra_genes <- dte_dtu_gene %>%
  dplyr::filter(!Gene %in% dte_dtu_tx$Gene)

sig_dtu %>%
  dplyr::filter(Gene %in% sig_tx$Gene,
                !Gene %in% sig_genes$Gene) %>%
  dplyr::filter(Gene %in% extra_genes$Gene) %>%
  dplyr::select("Gene",
                "Gene Name" = "gene_name",
                "Transcript",
                "Transcript Name" = "transcript_name",
                "DTU gene FDR" = "geneFDR",
                "DTU transcript FDR" = "txpFDR") %>%
  left_join(oxygenTranscripts %>%
              dplyr::select("Transcript", "DTE FDR" = "FDR"),
            by = "Transcript") %>%
  datatable()
```

Now check for genes that are DGE and DTU but not found in DTE. Basically just doing a full check here. What are the differences?
```{r DGE_DTU}
DGE_DTU <- sig_dtu %>%
  dplyr::filter(Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length()

sig_dtu %>%
  dplyr::filter(Gene %in% sig_genes$Gene,
                !Gene %in% sig_tx$Gene) %>%
  nrow() %>%
  print()
```

Now identify genes that overlap across all methods. 
```{r DGE_DTE_DTU}
DGE_DTE_DTU <- sig_dtu %>%
  dplyr::filter(Gene %in% sig_tx$Gene,
                Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length()

sig_dtu %>%
  dplyr::filter(Gene %in% sig_tx$Gene,
                Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length() %>%
  print()
```

Now lets grab genes in both DTE and DGE, but not in DTU
```{r DTE_DGE}
DTE_DGE <- sig_genes %>%
  dplyr::filter(Gene %in% sig_tx$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length()

sig_genes %>%
  dplyr::filter(Gene %in% sig_tx$Gene) %>%
  dplyr::select(Gene) %>%
  distinct() %>%
  as.matrix() %>%
  length() %>%
  print()
```

Now we get the total numbers for the area parameters, so we know how many genes were identified in each.
```{r DGE_only}
DGE_number <- sig_genes$Gene %>%
  unique() %>%
  length()

DTE_number <- sig_tx$Gene %>%
  unique() %>%
  length()

DTU_number <- sig_dtu$Gene %>%
  unique() %>%
  length()
```

The objects created in the previous code chunck are used here to plot the triple venn diagram which records the level of overlap between each of the different analyses run.
```{r draw_venn}
grid.newpage()

v <- draw.triple.venn(
  area1 = DGE_number,
  area2 = DTE_number,
  area3 = DTU_number,
  n12 = DTE_DGE,
  n13 = DGE_DTU,
  n23 = DTE_DTU,
  n123 = DGE_DTE_DTU,
  category = c(
    "DGE",
    "DTE",
    "DTU"),
  fill = c(
    "blue", 
    "red",
    "green"),
  lty = "blank",
  cex = 2,
  fontfamily = "sans",
  cat.fontfamily = "sans",
  cat.pos = c(200, 160, 360),
  cat.dist = c(0.05, 0.06, 0.075),
  cat.cex = 2,
  scaled = FALSE
)
```

Can add a title if you want, but not necessary for this R markdown
```{r eval=FALSE}
grid.arrange(
  gTree(children = v),
  top = textGrob("Overlap of significant genes from DGE, DTE, and DTU analysis",
                 gp = gpar(fontsize = 16,
                           font = 2),
                 vjust = 1.5)
)
```

# Greatest Differences Between Gene and Transcript Expression
In this section I compare the DGE and DTE results in order to find which transcripts are most genuinely masked by aggregation to the gene level. Sometimes transcripts can be masked by expression of a more "dominant" transcripts with multiple folds more expression than it, or a sister transcript that has an similar but opposite change in expression that neutralises any effect on the gene level, or if it has many other transcripts with similar changes in expression that are collapsed into a much seemingly smaller change on the gene level. Here we can take a look at whats happening in the data.
```{r DE_Differences}
DET_DEG_logFC_diff <- oxygenTranscripts %>%
  dplyr::select("Transcript",
                "Gene",
                "GeneName",
                "logCPM",
                "logFC",
                "FDR") %>%
  left_join(oxygenGenes %>%
              dplyr::select(
                "Gene",
                "gene_logFC" = "logFC",
                "geneFDR" = "FDR"
              ),
            by = "Gene") %>%
  left_join(transcript_names,
            by = c("Transcript" = "transcript_id")) %>%
  mutate(logFC_diff = (abs(logFC) - abs(gene_logFC))) %>%
  arrange(desc(logFC_diff))

DET_DEG_logFC_diff %>% dplyr::filter(logFC_diff > 1) %>%
  dplyr::filter(geneFDR > 0.05 & abs(gene_logFC) < 1) %>%
  dplyr::select(
    "Transcript ID" = "Transcript",
    "Transcript Name" = "transcript_name",
    "Gene Name" = "GeneName",
    "txp logFC" = "logFC",
    "logFC Diff" = "logFC_diff",
    "txp FDR" = "FDR",
    "gene FDR" = "geneFDR"
  )

significance_diff <- DET_DEG_logFC_diff %>% dplyr::filter(geneFDR > 0.05)

DET_DEG_logFC_diff %>%
  dplyr::filter(is.na(geneFDR)) %>%
  dplyr::select(
    "Transcript ID" = "Transcript",
    "Transcript Name" = "transcript_name",
    "Gene Name" = "GeneName",
    "logFC",
    "FDR"
  )

DT::datatable(DET_DEG_logFC_diff)
```

We should check to make sure the logFCs are not in the opposite direction between our gene and transcript counts. If differences in logFC cancel out then these results would be telling us the wrong thing
```{r}
sig_tx %>%
  dplyr::select(
    "Gene_txp" = "Gene",
    "logFC_txp" = "logFC") %>%
  left_join(sig_genes %>%
              dplyr::select(
                "Gene_gene" = "Gene",
                "logFC_gene" = "logFC"
              ),
            by = c("Gene_txp" = "Gene_gene")) %>%
  ggplot(
    aes(logFC_txp,
        logFC_gene)
    ) +
  geom_point() +
  xlab("logFC transcripts") +
  ylab("logFC genes") +
  theme_bw() +
  theme(axis.title = element_text(size = 20),
        axis.text = element_text(size = 20))
```

## Compare FDR and logFC between genes and transcripts
Lets see if there are any major differences between the FDR values for genes and transcripts, then extend this check to logFC
```{r compare}
gene_tx_diff <- DET_DEG_logFC_diff %>%
  dplyr::filter(logFC_diff > 1) %>%
  dplyr::filter(geneFDR > 0.05 & abs(gene_logFC) < 1) %>%
  dplyr::select(
    "Transcript",
    "TranscriptName" = "transcript_name",
    "Gene",
    "GeneName",
    "txp_logFC" = "logFC",
    "gene_logFC",
    "logFC_diff",
    "txpFDR" = "FDR",
    "geneFDR"
  ) 

DT::datatable(gene_tx_diff)
```


## Scale CPM for heatmap
Now we will plot out these differences. We will use a heatmap that can be made with the `pheatmap` package. I prefer this over manually making one with `ggplot2` as we can make nice annotation bars without needing to call `grid` or `cowplot`
We just want to make sure we have defined the order of our clusters in `clusterOrder` for visualisation purposes and appropriately scale our values to get the contrasting colours in the plot. Then lock everything up as a matrix
```{r scaledCPM}
## Scale by rows
# In this function we are calculating the mean and standard deviation for every row of the
# matrix. Then we subtract the mean from the raw value and divide by the calculated standard
# deviation to generate our scaled values. Dividing by the standard deviation means that

scale_rows <- function(x){
  m = apply(x, 1, mean, na.rm = T)
  s = apply(x, 1, sd, na.rm = T)
  return((x - m) / s)
}

clusterOrder <- dtelist %>%
  cpm() %>%
  as.data.frame() %>%
  rownames_to_column(var = "Transcript") %>%
  dplyr::filter(Transcript %in% gene_tx_diff$Transcript) %>%
  dplyr::arrange(Transcript) %>%
  set_rownames(gene_tx_diff %>%
                 dplyr::filter(!Transcript == "ENST00000372728") %>%
                 dplyr::arrange(Transcript) %>%
                 dplyr::select(Transcript) %>%
                 as.matrix() %>%
                 as.character()) %>%
  dplyr::select(-Transcript) %>%
  as.matrix() %>%
  dist() %>%
  hclust() %>%
  as.dendrogram(method = "ward.D2") %>%
  order.dendrogram()

scaled_cpm <- dtelist %>%
  cpm(log = TRUE) %>%
  as.data.frame() %>%
  rownames_to_column(var = "Transcript") %>%
  dplyr::filter(Transcript %in% gene_tx_diff$Transcript) %>%
  dplyr::arrange(Transcript) %>%
  set_rownames(gene_tx_diff %>%
                 dplyr::filter(!Transcript == "ENST00000372728") %>%
                 dplyr::arrange(Transcript) %>%
                 dplyr::select(Transcript) %>%
                 as.matrix() %>%
                 as.character()) %>%
  dplyr::select(-Transcript) %>%
  as.matrix()
```

## Plot pheatmap
With everything ready lets make this heatmap. We also want to define out annotation column. This should have rownames that correspond to the sample IDs in our matrix. This way, `pheatmap` will automatically know how to visualise the annotations 
```{r cluster_pheatmap}
clusteredMatrix <- scaled_cpm[clusterOrder, ]
clusteredMatrix <- clusteredMatrix[, (pd %>% 
                                        dplyr::filter(
                                          ID %in% colnames(clusteredMatrix)
                                        ) %>%
                                        dplyr::arrange(GestationalAge) %>%
                                        dplyr::select(ID) %>%
                                        as.matrix() %>%
                                        as.character)]

clusteredMatrix <- clusteredMatrix %>%
  as.data.frame() %>%
  rownames_to_column(var = "transcript_id") %>%
  left_join(transcript_names,
            by = "transcript_id") %>%
  dplyr::select(-transcript_id) %>%
  column_to_rownames("transcript_name") %>%
  as.matrix()

ann_col <- pd %>%
  dplyr::filter(ID %in% colnames(clusteredMatrix)) %>%
  dplyr::select(ID, "Gestation Group " = "Timepoint") %>%
  as.data.frame() %>%
  column_to_rownames("ID")

gene_row_order <- transcript_names %>%
  dplyr::filter(transcript_name %in% rownames(clusteredMatrix)) %>%
  left_join(oxygenTranscripts,
            by = c("transcript_id" = "Transcript")) %>%
  dplyr::select("Gene",
                "GeneName",
                "Transcript" = "transcript_id",
                "TranscriptName" = "transcript_name") %>%
  column_to_rownames("TranscriptName")

gene_row_order[clusterOrder, ]
```

And now we have the heatmap ready for visualisation
```{r}
pheatmap(clusteredMatrix,
         color = plasma(84),
         border_color = NA,
         show_colnames = FALSE,
         cluster_cols = FALSE,
         main = paste0(
           "Top 15 transcripts with greatest\n expression",
           " differences to the gene level"
         ),
         fontsize = 10,
         gaps_col = 27,
         cutree_rows = 4,
         clustering_method = "ward.D2",
         annotation_col = ann_col)
```

# Transcriptome profile
In this section I simply ask a few questions of the data. The purpose is so that I can offer a more in-depth understanding of what is within the data. This is kind of a data-driven approach, I'm not certain on what to find, and I am not testing anything. Just a bit of fun exploring to do

## Analysis Result Info
### How many significant genes across all analyses?
```{r all_genes}
all_genes <- c(sig_tx$Gene, sig_genes$Gene, sig_dtu$Gene) %>%
  unique()

print(length(all_genes))
```

### How many non coding in all 1642 genes?
```{r non_coding_genes}
grch38_gene_anno %>%
  dplyr::filter(gene_id %in% all_genes) %>%
  dplyr::count(gene_biotype, sort = TRUE) %>%
  print()
```

### What are the lncRNAs?
```{r lncRNAs}
lncRNAs <- grch38_gene_anno %>%
  dplyr::filter(gene_id %in% all_genes) %>%
  dplyr::filter(gene_biotype == "lncRNA")

sig_genes %>%
  dplyr::filter(Gene %in% lncRNAs$gene_id) %>%
  print()

sig_tx %>%
  dplyr::filter(Gene %in% lncRNAs$gene_id) %>%
  print()

sig_dtu %>%
  dplyr::filter(Gene %in% lncRNAs$gene_id) %>%
  print()
```

### What are all the significant transcripts between DTE and DTU?
```{r sig_txs}
all_tx <- c(sig_tx$Transcript, sig_dtu$Transcript) %>%
  unique()

print(length(all_tx))

all_tx_genes <- c(sig_tx$Gene, sig_dtu$Gene) %>%
  unique()

print(length(all_tx_genes))
```

### How many isoforms did these genes have?
```{r isoforms}
oxygenTranscripts %>% 
  dplyr::select("Gene",
                "Transcript") %>%
  rbind(oxygenDTU %>%
          dplyr::select("Gene",
                        "Transcript")) %>%
  dplyr::filter(Gene %in% all_tx_genes) %>%
  distinct() %>%
  dplyr::count(Gene, sort = TRUE) %>%
  dplyr::count(n, sort = TRUE, name = "Number of Genes") %>%
  dplyr::rename("Isoforms" = "n") %>%
  DT::datatable()
```

### What types of transcripts were they?
```{r transcript_biotypes}
grch38_txp_anno %>%
  dplyr::filter(transcript_id %in% all_tx) %>%
  dplyr::count(transcript_biotype, sort = TRUE) %>%
  DT::datatable()
```

## Info on the triple overlap genes
```{r}
triple_genes <- sig_dtu %>%
  dplyr::filter(Gene %in% sig_tx$Gene,
                Gene %in% sig_genes$Gene) %>%
  dplyr::select(Gene) %>%
  distinct()
```

What are the overlapped genes in DTE results?
```{r}
oxygenTranscripts %>%
  dplyr::filter(Gene %in% triple_genes$Gene) %>%
  dplyr::select("Gene",
                "GeneName",
                "Transcript",
                "TranscriptName",
                "logFC",
                "FDR") %>%
  dplyr::count(GeneName, sort = TRUE) %>%
  DT::datatable()
```

What are the overlapped genes in DTU results?
```{r}
oxygenDTU %>%
  dplyr::filter(Gene %in% triple_genes$Gene) %>%
  dplyr::select("Gene",
                "GeneName" = "gene_name",
                "Transcript",
                "TranscriptName" = "transcript_name",
                "geneFDR",
                "txpFDR") %>%
  dplyr::count(GeneName, sort = TRUE) %>%
  DT::datatable()
```

What are the significant overlapped genes in DTE results?
```{r}
oxygenTranscripts %>%
  dplyr::filter(Gene %in% triple_genes$Gene) %>%
  dplyr::select("Gene",
                "GeneName",
                "Transcript",
                "TranscriptName",
                "logFC",
                "FDR") %>%
  dplyr::filter(FDR < 0.05 & abs(logFC) > 1) %>%
  dplyr::count(GeneName, sort = TRUE) %>%
  DT::datatable()
```

What are the significant overlapped genes in DTU results?
```{r}
oxygenDTU %>%
  dplyr::filter(Gene %in% triple_genes$Gene) %>%
  dplyr::select("Gene",
                "GeneName" = "gene_name",
                "Transcript",
                "TranscriptName" = "transcript_name",
                "geneFDR",
                "txpFDR") %>%
  dplyr::filter(geneFDR < 0.05 & txpFDR < 0.05) %>%
  dplyr::count(GeneName, sort = TRUE) %>%
  dplyr::filter(n > 1) %>%
  DT::datatable()
```

The same but which ones appear only once?
```{r}
oxygenDTU %>%
  dplyr::filter(Gene %in% triple_genes$Gene) %>%
  dplyr::select("Gene",
                "GeneName" = "gene_name",
                "Transcript",
                "TranscriptName" = "transcript_name",
                "geneFDR",
                "txpFDR") %>%
  dplyr::filter(geneFDR < 0.05 & txpFDR < 0.05) %>%
  dplyr::count(GeneName, sort = TRUE) %>%
  dplyr::filter(n == 1) %>%
  DT::datatable()
```

## Transcript specific information
### What are their biotypes?
```{r}
oxt_tx_triple <- oxygenTranscripts %>%
  dplyr::filter(Gene %in% triple_genes$Gene)

oxy_dtu_triple <- oxygenDTU %>% 
  dplyr::filter(Gene %in% triple_genes$Gene)
```

Check to see we have complete representation of all transcripts here and then count the transcript biotypes
```{r}
all(oxy_dtu_triple$Transcript %in% oxt_tx_triple$Transcript)

oxt_tx_triple %>%
  dplyr::count(Transcript_biotype) %>%
  DT::datatable()
```

### What transcripts have retained introns?
```{r}
oxt_tx_triple %>% 
  dplyr::filter(Transcript_biotype == "retained_intron") %>%
  DT::datatable()
```

### What transcripts have untranslated transcripts?
```{r}
oxt_tx_triple %>%
  dplyr::filter(Transcript_biotype == "processed_transcript") %>%
  DT::datatable()
```

### Which transcripts are targeted by nonsense mediated decay?
```{r}
oxt_tx_triple %>%
  dplyr::filter(Transcript_biotype == "nonsense_mediated_decay") %>%
  DT::datatable()
```

## Isoform information in genes
### How many isoforms transcribed by each gene?
```{r}
sig_tx %>%
  dplyr::count(Gene, sort = TRUE) %>%
  DT::datatable()

sig_tx %>% 
  dplyr::count(Gene,
               sort = TRUE, 
               name = "isoforms") %>%
  dplyr::count(isoforms,
               sort = TRUE) %>%
  DT::datatable()
```

### Which genes had the most amount of significantly DE isoforms?
```{r}
sig_tx %>%
  dplyr::count(Gene, sort = TRUE) %>%
  left_join(sig_tx %>%
              dplyr::select("Gene", "GeneName") %>%
              distinct(),
            by = "Gene") %>%
  DT::datatable()
```

## Heatmap transcripts information
```{r}
oxygenTranscripts %>%
  dplyr::filter(TranscriptName %in% rownames(clusteredMatrix)) %>%
  DT::datatable()
```

